// @traceloop/instrumentation-utils/dist/index.js
'use strict';

var incubating = require('@opentelemetry/semantic-conventions/incubating');

/**
 * Utility functions for formatting LLM messages according to OpenTelemetry
 * Semantic Conventions v1.40.0 for gen_ai attributes.
 *
 * These formatters produce JSON strings for attributes like:
 * - gen_ai.system_instructions
 * - gen_ai.input.messages
 * - gen_ai.output.messages
 *
 * JSON schemas sourced from:
 * - https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-system-instructions.json
 * - https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-input-messages.json
 * - https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-output-messages.json
 */
// =============================================================================
// OTel Part Types (shared across all three schemas)
// =============================================================================
//
// TextPart:              { type: "text",                  content: string }
// ToolCallRequestPart:   { type: "tool_call",             id?, name, arguments? }
// ToolCallResponsePart:  { type: "tool_call_response",    id?, response }
// ServerToolCallPart:    { type: "server_tool_call",      id?, name, server_tool_call }
// ServerToolCallRespPart:{ type: "server_tool_call_response", id?, server_tool_call_response }
// BlobPart:              { type: "blob",                  modality, mime_type?, content (base64) }
// UriPart:               { type: "uri",                   modality, mime_type?, uri }
// FilePart:              { type: "file",                  modality, mime_type?, file_id }
// ReasoningPart:         { type: "reasoning",             content: string }
// GenericPart:           { type: string, ...any }          — extensibility escape hatch
//
// Modality enum: "image" | "video" | "audio"
// Role enum:     "system" | "user" | "assistant" | "tool"
// =============================================================================
// formatSystemInstructions
// =============================================================================
//
// OTel contract — gen_ai.system_instructions (array of parts, NO wrapping message):
//
// [
//   TextPart            | { type: "text",               content: string }
//   ToolCallRequestPart | { type: "tool_call",          id?, name, arguments? }
//   ToolCallResponsePart| { type: "tool_call_response", id?, response }
//   BlobPart            | { type: "blob",               modality, mime_type?, content }
//   FilePart            | { type: "file",               modality, mime_type?, file_id }
//   UriPart             | { type: "uri",                modality, mime_type?, uri }
//   ReasoningPart       | { type: "reasoning",          content: string }
//   GenericPart         | { type: string, ...any }
// ]
//
// Example (string input):
//   [ { "type": "text", "content": "You are a helpful assistant." } ]
//
// Example (array input with mixed blocks):
//   [
//     { "type": "text",  "content": "You are a helpful assistant." },
//     { "type": "text",  "content": "Always respond in English." }
//   ]
/**
 * Formats system instructions for gen_ai.system_instructions attribute.
 *
 * @param system - System instruction as a plain string or an array of content blocks
 * @returns JSON string — a flat array of OTel part objects (no wrapping message)
 */
function formatSystemInstructions(system) {
    if (typeof system === "string") {
        return JSON.stringify([{ type: "text", content: system }]);
    }
    // Each block in the array maps directly to its OTel part — no wrapper object
    return JSON.stringify(system.map((block) => {
        if (block.type === "text") {
            return { type: "text", content: block.text };
        }
        // Preserve any other block type as a GenericPart
        return Object.assign({}, block);
    }));
}
// =============================================================================
// formatInputMessages
// =============================================================================
//
// OTel contract — gen_ai.input.messages (array of ChatMessage objects):
//
// [
//   {
//     role: "system" | "user" | "assistant" | "tool" | string,
//     parts: [
//       TextPart            | { type: "text",                        content: string }
//       ToolCallRequestPart | { type: "tool_call",                   id?, name, arguments? }
//       ToolCallResponsePart| { type: "tool_call_response",          id?, response }
//       ServerToolCallPart  | { type: "server_tool_call",            id?, name, server_tool_call }
//       ServerToolCallResp  | { type: "server_tool_call_response",   id?, server_tool_call_response }
//       BlobPart            | { type: "blob",                        modality, mime_type?, content }
//       FilePart            | { type: "file",                        modality, mime_type?, file_id }
//       UriPart             | { type: "uri",                         modality, mime_type?, uri }
//       ReasoningPart       | { type: "reasoning",                   content: string }
//       GenericPart         | { type: string, ...any }
//     ],
//     name?: string
//   }
// ]
//
// Example (text only):
//   [
//     { "role": "user", "parts": [ { "type": "text", "content": "Hello!" } ] }
//   ]
//
// Example (multimodal with image):
//   [
//     {
//       "role": "user",
//       "parts": [
//         { "type": "text",  "content": "What is in this image?" },
//         { "type": "blob",  "modality": "image", "mime_type": "image/png", "content": "<base64>" }
//       ]
//     }
//   ]
//
// Example (with tool call):
//   [
//     {
//       "role": "assistant",
//       "parts": [ { "type": "tool_call", "id": "call_123", "name": "get_weather", "arguments": { "city": "Paris" } } ]
//     },
//     {
//       "role": "tool",
//       "parts": [ { "type": "tool_call_response", "id": "call_123", "response": "rainy, 57°F" } ]
//     }
//   ]
/**
 * Formats chat messages for gen_ai.input.messages attribute.
 *
 * @param messages - Array of message objects with role and content
 * @param contentBlockMapper - Provider-specific function to map a content block to an OTel part
 * @returns JSON string — array of ChatMessage objects with role and typed parts
 */
function formatInputMessages(messages, contentBlockMapper, 
// When false, message content is omitted from the span (privacy/security).
// Role is always kept — it is metadata, not content.
traceContent = true) {
    return JSON.stringify(messages.map((message) => ({
        role: message.role,
        parts: traceContent
            ? typeof message.content === "string"
                ? [{ type: "text", content: message.content }]
                : message.content.map(contentBlockMapper)
            : [],
    })));
}
// =============================================================================
// formatInputMessagesFromPrompt
// =============================================================================
//
// OTel contract — same as gen_ai.input.messages above.
// Convenience wrapper for a plain text_completion prompt with a single user turn.
//
// Example output:
//   [ { "role": "user", "parts": [ { "type": "text", "content": "Summarise this document." } ] } ]
/**
 * Formats a text completion prompt for gen_ai.input.messages attribute.
 *
 * @param prompt - The text prompt string
 * @returns JSON string — array with a single user ChatMessage
 */
function formatInputMessagesFromPrompt(prompt, 
// When false, message content is omitted from the span (privacy/security).
// Role is always kept — it is metadata, not content.
traceContent = true) {
    return JSON.stringify([
        {
            role: "user",
            parts: traceContent ? [{ type: "text", content: prompt }] : [],
        },
    ]);
}
// =============================================================================
// formatOutputMessage
// =============================================================================
//
// OTel contract — gen_ai.output.messages (array of OutputMessage objects):
//
// [
//   {
//     role: "assistant" | string,
//     finish_reason: string | null,
//     parts: [
//       TextPart            | { type: "text",                        content: string }
//       ToolCallRequestPart | { type: "tool_call",                   id?, name, arguments? }
//       ToolCallResponsePart| { type: "tool_call_response",          id?,response }
//       ServerToolCallPart  | { type: "server_tool_call",            id?, name, server_tool_call }
//       ServerToolCallResp  | { type: "server_tool_call_response",   id?, server_tool_call_response }
//       BlobPart            | { type: "blob",                        modality, mime_type?, content }
//       FilePart            | { type: "file",                        modality, mime_type?, file_id }
//       UriPart             | { type: "uri",                         modality, mime_type?, uri }
//       ReasoningPart       | { type: "reasoning",                   content: string }
//       GenericPart         | { type: string, ...any }
//     ]
//   }
// ]
//
// Example (simple text response):
//   [
//     {
//       "role": "assistant",
//       "finish_reason": "stop",
//       "parts": [ { "type": "text", "content": "The capital of France is Paris." } ]
//     }
//   ]
//
// Example (tool call response):
//   [
//     {
//       "role": "assistant",
//       "finish_reason": "tool_call",
//       "parts": [ { "type": "tool_call", "id": "call_123", "name": "get_weather", "arguments": { "city": "Paris" } } ]
//     }
//   ]
//
// Example (with reasoning/thinking block):
//   [
//     {
//       "role": "assistant",
//       "finish_reason": "stop",
//       "parts": [
//         { "type": "reasoning", "content": "Let me think about this..." },
//         { "type": "text",      "content": "The answer is 42." }
//       ]
//     }
//   ]
/**
 * Formats output content for gen_ai.output.messages attribute.
 *
 * @param content - Array of content blocks or string completion
 * @param stopReason - The stop reason from the LLM response
 * @param finishReasonMap - Mapping of provider-specific stop reasons to standard finish reasons
 * @param type - Type of completion: "chat", "generate_content", or "text_completion" (use GEN_AI_OPERATION_NAME_VALUE_* constants)
 * @param contentBlockMapper - Provider-specific function to map a content block to an OTel part
 * @returns JSON string — array with a single assistant OutputMessage
 */
function formatOutputMessage(content, stopReason, finishReasonMap, type, contentBlockMapper, 
// When false, message content is omitted from the span (privacy/security).
// Role and finish_reason are always kept — they are metadata, not content.
traceContent = true) {
    var _a;
    const outputMessage = {
        role: "assistant",
        // Always include finish_reason. Use "" when the model did not return a stop reason.
        finish_reason: stopReason
            ? ((_a = finishReasonMap[stopReason]) !== null && _a !== void 0 ? _a : stopReason)
            : "",
        parts: [],
    };
    if (traceContent) {
        if ((type === incubating.GEN_AI_OPERATION_NAME_VALUE_CHAT ||
            type === incubating.GEN_AI_OPERATION_NAME_VALUE_GENERATE_CONTENT) &&
            Array.isArray(content)) {
            outputMessage.parts = content.map(contentBlockMapper);
        }
        else if (type === incubating.GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION &&
            typeof content === "string") {
            outputMessage.parts = [{ type: "text", content }];
        }
    }
    return JSON.stringify([outputMessage]);
}

/**
 * Provider-specific content block mappers for OTel gen_ai attributes.
 *
 * Each function maps a single provider SDK content block to its
 * OTel-compliant part object, as defined in the gen_ai semantic conventions.
 *
 * These are passed as the `contentBlockMapper` parameter to
 * `formatInputMessages` and `formatOutputMessage`.
 */
// =============================================================================
// Anthropic
// =============================================================================
//  Maps a single Anthropic SDK content block to its OTel-compliant part object.
//
//   text                      → TextPart
//   image (base64)            → BlobPart       { modality: "image", mime_type from block }
//   image (url)               → UriPart        { modality: "image" }
//   document (base64)         → BlobPart       { mime_type from block — may be "application/pdf" or "text/plain" }
//   document (text)           → TextPart
//   document (url)            → UriPart        { mime_type: "application/pdf" }
//   document (file)           → FilePart       { file_id from source.file_id }
//   tool_use                  → ToolCallRequestPart
//   tool_result               → ToolCallResponsePart  (response may be string or content block array)
//   server_tool_use           → ServerToolCallPart
//   thinking                  → ReasoningPart  (OTel type is "reasoning", not "thinking")
//   redacted_thinking         → GenericPart    { type: "redacted_thinking" }  — data intentionally omitted
//   <unknown>                 → GenericPart
/**
 * Maps a single Anthropic content block to its OTel-compliant part object.
 * Used by formatInputMessages, formatOutputMessage, and formatSystemInstructions.
 */
function mapAnthropicContentBlock(block) {
    if (typeof block === "string") {
        return { type: "text", content: block };
    }
    switch (block.type) {
        // -------------------------------------------------------------------------
        // Text
        // -------------------------------------------------------------------------
        case "text":
            return { type: "text", content: block.text };
        // -------------------------------------------------------------------------
        // Image — sources: base64 | url
        // -------------------------------------------------------------------------
        case "image": {
            const src = block.source;
            if ((src === null || src === void 0 ? void 0 : src.type) === "base64") {
                return {
                    type: "blob",
                    modality: "image",
                    mime_type: src.media_type, // e.g. "image/jpeg", "image/png", "image/webp", "image/gif"
                    content: src.data,
                };
            }
            if ((src === null || src === void 0 ? void 0 : src.type) === "url") {
                return { type: "uri", modality: "image", uri: src.url };
            }
            // Unknown image source shape — fall through to GenericPart
            return Object.assign({ type: block.type }, block);
        }
        // -------------------------------------------------------------------------
        // Document — sources: base64 | text | url | file
        // OTel modality accepts any string (anyOf: [Modality enum, string]), so
        // "document" is a valid value for BlobPart/UriPart/FilePart modality.
        // -------------------------------------------------------------------------
        case "document": {
            const src = block.source;
            switch (src === null || src === void 0 ? void 0 : src.type) {
                case "base64":
                    // Inline base64 — media_type comes from the block itself, never hardcode it
                    // (can be "application/pdf" or "text/plain")
                    return {
                        type: "blob",
                        modality: "document",
                        mime_type: src.media_type,
                        content: src.data,
                    };
                case "text":
                    // Inline plain text document
                    return { type: "text", content: src.data };
                case "url":
                    // URL-referenced PDF
                    return {
                        type: "uri",
                        modality: "document",
                        mime_type: "application/pdf",
                        uri: src.url,
                    };
                case "file":
                    // Files API reference — file_id is on source.file_id
                    return { type: "file", modality: "document", file_id: src.file_id };
                default:
                    return Object.assign({ type: block.type }, block);
            }
        }
        // -------------------------------------------------------------------------
        // Tool use (client-side function call requested by the model)
        // -------------------------------------------------------------------------
        case "tool_use":
            return {
                type: "tool_call",
                id: block.id,
                name: block.name,
                arguments: block.input,
            };
        // -------------------------------------------------------------------------
        // Tool result (client sends function result back to the model)
        // Note: block.content may be a string OR an array of content blocks
        // -------------------------------------------------------------------------
        case "tool_result":
            return {
                type: "tool_call_response",
                id: block.tool_use_id,
                response: block.content,
            };
        // -------------------------------------------------------------------------
        // Server-side tool use (e.g. web_search, code_execution — run by Anthropic)
        // -------------------------------------------------------------------------
        case "server_tool_use":
            return {
                type: "server_tool_call",
                id: block.id,
                name: block.name,
                server_tool_call: Object.assign({ type: block.name }, block.input),
            };
        // -------------------------------------------------------------------------
        // Extended thinking / reasoning
        // -------------------------------------------------------------------------
        case "thinking":
            // OTel ReasoningPart uses type "reasoning", not "thinking"
            return { type: "reasoning", content: block.thinking };
        case "redacted_thinking":
            // Anthropic redacts thinking blocks when they contain sensitive content.
            // The `data` field is an opaque encrypted blob — intentionally not logged.
            return { type: "redacted_thinking" };
        // -------------------------------------------------------------------------
        // Unknown / future block types — preserve as GenericPart
        // -------------------------------------------------------------------------
        default:
            return Object.assign({ type: block.type }, block);
    }
}
// =============================================================================
// OpenAI
// =============================================================================
//  Maps a single OpenAI SDK input content part to its OTel-compliant part object.
//
//  This mapper handles the REQUEST-side content parts that appear in user messages
//  (ChatCompletionContentPart union). OpenAI's response is flat (not block-based),
//  so output mapping is handled by buildOpenAIOutputMessage in message-helpers.ts.
//
//   text                      → TextPart
//   image_url (regular URL)   → UriPart        { modality: "image" }
//   image_url (data: URI)     → BlobPart       { modality: "image", mime_type parsed from URI }
//   input_audio               → BlobPart       { modality: "audio", mime_type: "audio/{format}" }
//   file (file_id)            → FilePart       { modality: "document", file_id }
//   file (file_data)          → BlobPart       { modality: "document", mime_type }
//   refusal                   → GenericPart    { type: "refusal", content }
//   <unknown>                 → GenericPart
/**
 * Maps a single OpenAI SDK content part to its OTel-compliant part object.
 * Used for input content parts in user messages.
 */
function mapOpenAIContentBlock(block) {
    var _a, _b, _c, _d, _e;
    if (typeof block === "string") {
        return { type: "text", content: block };
    }
    switch (block.type) {
        // -------------------------------------------------------------------------
        // Text
        // -------------------------------------------------------------------------
        case "text":
            return { type: "text", content: block.text };
        // -------------------------------------------------------------------------
        // Image URL — can be a regular URL or a base64 data: URI
        // -------------------------------------------------------------------------
        case "image_url": {
            const url = (_a = block.image_url) === null || _a === void 0 ? void 0 : _a.url;
            if (url && url.startsWith("data:")) {
                // Parse data URI: data:image/png;base64,xxxx
                const match = url.match(/^data:([^;]+);base64,(.+)$/);
                if (match) {
                    return {
                        type: "blob",
                        modality: "image",
                        mime_type: match[1],
                        content: match[2],
                    };
                }
            }
            return { type: "uri", modality: "image", uri: url };
        }
        // -------------------------------------------------------------------------
        // Input audio — base64 encoded audio data with format
        // -------------------------------------------------------------------------
        case "input_audio":
            return {
                type: "blob",
                modality: "audio",
                mime_type: `audio/${((_b = block.input_audio) === null || _b === void 0 ? void 0 : _b.format) || "wav"}`,
                content: (_c = block.input_audio) === null || _c === void 0 ? void 0 : _c.data,
            };
        // -------------------------------------------------------------------------
        // File — can be file_id reference or inline file_data
        // OpenAI file content parts are used for documents (PDFs etc.), not images
        // (images use image_url). modality accepts any string, so "document" is valid.
        // -------------------------------------------------------------------------
        case "file": {
            if ((_d = block.file) === null || _d === void 0 ? void 0 : _d.file_id) {
                return Object.assign({ type: "file", modality: "document", file_id: block.file.file_id }, (block.file.filename && { filename: block.file.filename }));
            }
            if ((_e = block.file) === null || _e === void 0 ? void 0 : _e.file_data) {
                // Inline file data → BlobPart. OTel FilePart is a reference type (file_id only).
                return {
                    type: "blob",
                    modality: "document",
                    mime_type: block.file.mime_type || "application/octet-stream",
                    content: block.file.file_data,
                };
            }
            return Object.assign({ type: block.type }, block);
        }
        // -------------------------------------------------------------------------
        // Refusal — assistant content part indicating safety refusal
        // -------------------------------------------------------------------------
        case "refusal":
            return { type: "refusal", content: block.refusal };
        // -------------------------------------------------------------------------
        // Unknown / future content part types — preserve as GenericPart
        // -------------------------------------------------------------------------
        default:
            return Object.assign({ type: block.type }, block);
    }
}
// =============================================================================
// Bedrock
// =============================================================================
//  Maps a single Bedrock SDK content block to its OTel-compliant part object.
//
//  Bedrock is used by multiple providers (Anthropic, AI21, Amazon Nova, Meta,
//  Cohere) — all share the same block schema on the Bedrock layer:
//
//   plain string              → TextPart
//   { type: "text", text }    → TextPart
//   { text } (Nova format)    → TextPart
//   tool_use                  → ToolCallRequestPart
//   tool_result               → ToolCallResponsePart
//   <unknown>                 → GenericPart
/** Bedrock SDK content block type strings. */
const BedrockBlockType = {
    TEXT: "text",
    TOOL_USE: "tool_use",
    TOOL_RESULT: "tool_result",
};
/** OTel gen_ai part type strings used by the Bedrock mapper. */
const BedrockOtelPartType = {
    TEXT: "text",
    TOOL_CALL: "tool_call",
    TOOL_CALL_RESPONSE: "tool_call_response",
};
/**
 * Maps a single Bedrock content block to its OTel-compliant part object.
 * Used by formatInputMessages and formatOutputMessage for all Bedrock providers.
 */
const mapBedrockContentBlock = (block) => {
    if (typeof block === "string") {
        return { type: BedrockOtelPartType.TEXT, content: block };
    }
    switch (block.type) {
        case BedrockBlockType.TEXT:
            return { type: BedrockOtelPartType.TEXT, content: block.text };
        case BedrockBlockType.TOOL_USE:
            return {
                type: BedrockOtelPartType.TOOL_CALL,
                id: block.id,
                name: block.name,
                arguments: block.input,
            };
        case BedrockBlockType.TOOL_RESULT:
            return {
                type: BedrockOtelPartType.TOOL_CALL_RESPONSE,
                id: block.tool_use_id,
                response: block.content,
            };
        default:
            if (!block.type && block.text !== undefined) {
                return { type: BedrockOtelPartType.TEXT, content: block.text };
            }
            return Object.assign({ type: block.type }, block);
    }
};
/**
 * Maps a single Vercel AI SDK content part to its OTel-compliant part object.
 *
 * Field names follow AI SDK v6:
 *   ToolCallPart:   toolCallId, toolName, input
 *   ToolResultPart: toolCallId, toolName, output (ToolResultOutput union)
 *   ImagePart:      image (DataContent | URL), optional mediaType
 *   FilePart:       data (DataContent | URL), mediaType (required)
 *   ReasoningPart:  text
 */
const mapAiSdkContentPart = (part) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    if (!part || typeof part !== "object") {
        return { type: "text" /* AiSdkPartType.Text */, content: String(part !== null && part !== void 0 ? part : "") };
    }
    switch (part.type) {
        case "text" /* AiSdkPartType.Text */:
            return { type: "text" /* AiSdkPartType.Text */, content: (_a = part.text) !== null && _a !== void 0 ? _a : "" };
        // OTel type is "reasoning", AI SDK v6 field is `text`
        case "reasoning" /* AiSdkPartType.Reasoning */:
            return { type: "reasoning" /* AiSdkPartType.Reasoning */, content: (_b = part.text) !== null && _b !== void 0 ? _b : "" };
        case "tool-call" /* AiSdkPartType.ToolCall */:
            // AI SDK v6: toolCallId, toolName, input
            return {
                type: "tool_call",
                id: (_c = part.toolCallId) !== null && _c !== void 0 ? _c : null,
                name: part.toolName,
                arguments: part.input,
            };
        case "tool-result" /* AiSdkPartType.ToolResult */: {
            // AI SDK v6: output is ToolResultOutput — { type: 'text'|'json', value } union
            // Unwrap to the actual value for tracing; fall back to the full object if unknown shape
            const output = part.output;
            const response = output && typeof output === "object" && "value" in output
                ? output.value
                : output;
            return {
                type: "tool_call_response",
                id: (_d = part.toolCallId) !== null && _d !== void 0 ? _d : null,
                response,
            };
        }
        case "image" /* AiSdkPartType.Image */: {
            // AI SDK v6: image is DataContent | URL; optional mediaType
            const imgSrc = (_e = part.image) !== null && _e !== void 0 ? _e : null;
            const mimeType = (_f = part.mediaType) !== null && _f !== void 0 ? _f : null;
            if (imgSrc instanceof URL) {
                return {
                    type: "uri",
                    modality: "image",
                    uri: imgSrc.href,
                    mime_type: mimeType,
                };
            }
            if (typeof imgSrc === "string") {
                if (imgSrc.startsWith("data:")) {
                    const [header, data] = imgSrc.split(",");
                    const detectedMime = (_h = (_g = header.match(/data:([^;]+)/)) === null || _g === void 0 ? void 0 : _g[1]) !== null && _h !== void 0 ? _h : mimeType;
                    return Object.assign(Object.assign({ type: "blob", modality: "image" }, (detectedMime ? { mime_type: detectedMime } : {})), { content: data || "" });
                }
                return {
                    type: "uri",
                    modality: "image",
                    uri: imgSrc,
                    mime_type: mimeType,
                };
            }
            if (imgSrc != null) {
                // Binary data (Uint8Array / ArrayBuffer / Buffer) — base64 encode
                const bytes = imgSrc instanceof ArrayBuffer ? new Uint8Array(imgSrc) : imgSrc;
                const b64 = Buffer.from(bytes).toString("base64");
                return {
                    type: "blob",
                    modality: "image",
                    mime_type: mimeType,
                    content: b64,
                };
            }
            return { type: "blob", modality: "image", content: "" };
        }
        case "file" /* AiSdkPartType.File */: {
            // AI SDK v6: data is DataContent | URL; mediaType is required
            const fileSrc = (_j = part.data) !== null && _j !== void 0 ? _j : null;
            const mimeType = (_k = part.mediaType) !== null && _k !== void 0 ? _k : null;
            // Infer modality from MIME type (best-effort)
            const modality = (mimeType === null || mimeType === void 0 ? void 0 : mimeType.startsWith("image/"))
                ? "image"
                : (mimeType === null || mimeType === void 0 ? void 0 : mimeType.startsWith("audio/"))
                    ? "audio"
                    : (mimeType === null || mimeType === void 0 ? void 0 : mimeType.startsWith("video/"))
                        ? "video"
                        : "document";
            if (fileSrc instanceof URL) {
                return {
                    type: "uri",
                    modality,
                    uri: fileSrc.href,
                    mime_type: mimeType,
                };
            }
            if (typeof fileSrc === "string") {
                if (fileSrc.startsWith("data:")) {
                    const [, data] = fileSrc.split(",");
                    return {
                        type: "blob",
                        modality,
                        mime_type: mimeType,
                        content: data || "",
                    };
                }
                return { type: "uri", modality, uri: fileSrc, mime_type: mimeType };
            }
            if (fileSrc != null) {
                const bytes = fileSrc instanceof ArrayBuffer ? new Uint8Array(fileSrc) : fileSrc;
                const b64 = Buffer.from(bytes).toString("base64");
                return { type: "blob", modality, mime_type: mimeType, content: b64 };
            }
            return { type: "blob", modality, content: "" };
        }
        default:
            // GenericPart — preserve unknown types as-is
            return Object.assign({ type: part.type }, part);
    }
};
/**
 * Converts Vercel AI SDK message content to an array of OTel parts.
 * Accepts: plain string, array of SDK content parts, or a JSON-serialized array
 * (the AI SDK serializes content arrays as JSON strings in span attributes).
 */
const mapAiSdkMessageContent = (content) => {
    if (typeof content === "string") {
        try {
            const parsed = JSON.parse(content);
            if (Array.isArray(parsed)) {
                return parsed.map(mapAiSdkContentPart);
            }
        }
        catch (_a) {
            // plain string
        }
        return [{ type: "text", content }];
    }
    if (Array.isArray(content)) {
        return content.map(mapAiSdkContentPart);
    }
    if (content && typeof content === "object") {
        return [mapAiSdkContentPart(content)];
    }
    return [{ type: "text", content: String(content !== null && content !== void 0 ? content : "") }];
};

exports.formatInputMessages = formatInputMessages;
exports.formatInputMessagesFromPrompt = formatInputMessagesFromPrompt;
exports.formatOutputMessage = formatOutputMessage;
exports.formatSystemInstructions = formatSystemInstructions;
exports.mapAiSdkContentPart = mapAiSdkContentPart;
exports.mapAiSdkMessageContent = mapAiSdkMessageContent;
exports.mapAnthropicContentBlock = mapAnthropicContentBlock;
exports.mapBedrockContentBlock = mapBedrockContentBlock;
exports.mapOpenAIContentBlock = mapOpenAIContentBlock;
