// @traceloop/instrumentation-langchain/dist/index.js
'use strict';

var api = require('@opentelemetry/api');
var instrumentation = require('@opentelemetry/instrumentation');
var aiSemanticConventions = require('@traceloop/ai-semantic-conventions');
var tslib = require('tslib');
var base = require('@langchain/core/callbacks/base');
var incubating = require('@opentelemetry/semantic-conventions/incubating');

/*
 * Copyright Traceloop
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Combined finish reason map covering all providers LangChain wraps.
// Unknown reasons pass through unchanged via the `?? reason` fallback.
//
// NOTE FOR REVIEWERS: This is intentionally a single flat map as a stopgap.
// A follow-up PR (`feat/centralize-finish-reason-maps`) will move all per-provider
// finish reason maps (Anthropic, OpenAI, and this one) into @traceloop/instrumentation-utils
// so they can be shared across all instrumentations. At that point this map will be
// replaced with per-provider imports from that package.
const langchainFinishReasonMap = {
    // OpenAI / Azure / Together
    stop: aiSemanticConventions.FinishReasons.STOP,
    length: aiSemanticConventions.FinishReasons.LENGTH,
    tool_calls: aiSemanticConventions.FinishReasons.TOOL_CALL, // plural → singular
    content_filter: aiSemanticConventions.FinishReasons.CONTENT_FILTER,
    function_call: aiSemanticConventions.FinishReasons.TOOL_CALL, // deprecated
    // Anthropic / Bedrock (Anthropic models)
    end_turn: aiSemanticConventions.FinishReasons.STOP,
    stop_sequence: aiSemanticConventions.FinishReasons.STOP,
    tool_use: aiSemanticConventions.FinishReasons.TOOL_CALL,
    max_tokens: aiSemanticConventions.FinishReasons.LENGTH,
    // Google (Vertex AI / Gemini)
    STOP: aiSemanticConventions.FinishReasons.STOP,
    MAX_TOKENS: aiSemanticConventions.FinishReasons.LENGTH,
    SAFETY: aiSemanticConventions.FinishReasons.CONTENT_FILTER,
    RECITATION: aiSemanticConventions.FinishReasons.CONTENT_FILTER,
    // Cohere
    COMPLETE: aiSemanticConventions.FinishReasons.STOP,
    ERROR_TOXIC: aiSemanticConventions.FinishReasons.CONTENT_FILTER,
};
class TraceloopCallbackHandler extends base.BaseCallbackHandler {
    constructor(tracer, traceContent = true) {
        super();
        this.name = "traceloop_callback_handler";
        this.spans = new Map();
        this.tracer = tracer;
        this.traceContent = traceContent;
    }
    handleChatModelStart(llm, messages, runId, _parentRunId, _extraParams, _tags, _metadata, _runName) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            const className = ((_a = llm.id) === null || _a === void 0 ? void 0 : _a[llm.id.length - 1]) || "unknown";
            const vendor = this.detectVendor(llm);
            const operationType = incubating.GEN_AI_OPERATION_NAME_VALUE_CHAT;
            const requestModel = this.extractModelFromExtraParams(_extraParams);
            const spanName = requestModel
                ? `${operationType} ${requestModel}`
                : `${operationType} ${className}`;
            const span = this.tracer.startSpan(spanName, {
                kind: api.SpanKind.CLIENT,
            });
            const flatMessages = messages.flat();
            span.setAttributes({
                [incubating.ATTR_GEN_AI_PROVIDER_NAME]: vendor,
                [incubating.ATTR_GEN_AI_OPERATION_NAME]: operationType,
            });
            if (requestModel) {
                span.setAttribute(incubating.ATTR_GEN_AI_REQUEST_MODEL, requestModel);
            }
            if (this.traceContent && flatMessages.length > 0) {
                const inputMessages = flatMessages.map((message) => {
                    const role = this.mapMessageTypeToRole(message.type);
                    const content = typeof message.content === "string"
                        ? message.content
                        : JSON.stringify(message.content);
                    return {
                        role,
                        parts: [{ type: "text", content }],
                    };
                });
                span.setAttribute(incubating.ATTR_GEN_AI_INPUT_MESSAGES, JSON.stringify(inputMessages));
            }
            this.spans.set(runId, { span, runId, operationType, requestModel });
        });
    }
    handleLLMStart(llm, prompts, runId, _parentRunId, _extraParams, _tags, _metadata, _runName) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            const className = ((_a = llm.id) === null || _a === void 0 ? void 0 : _a[llm.id.length - 1]) || "unknown";
            const vendor = this.detectVendor(llm);
            const operationType = incubating.GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION;
            const requestModel = this.extractModelFromExtraParams(_extraParams);
            const spanName = requestModel
                ? `${operationType} ${requestModel}`
                : `${operationType} ${className}`;
            const span = this.tracer.startSpan(spanName, {
                kind: api.SpanKind.CLIENT,
            });
            span.setAttributes({
                [incubating.ATTR_GEN_AI_PROVIDER_NAME]: vendor,
                [incubating.ATTR_GEN_AI_OPERATION_NAME]: operationType,
            });
            if (requestModel) {
                span.setAttribute(incubating.ATTR_GEN_AI_REQUEST_MODEL, requestModel);
            }
            if (this.traceContent && prompts.length > 0) {
                const inputMessages = prompts.map((prompt) => ({
                    role: "user",
                    parts: [{ type: "text", content: prompt }],
                }));
                span.setAttribute(incubating.ATTR_GEN_AI_INPUT_MESSAGES, JSON.stringify(inputMessages));
            }
            this.spans.set(runId, { span, runId, operationType, requestModel });
        });
    }
    handleLLMEnd(output, runId, _parentRunId, _tags, _extraParams) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f;
            const spanData = this.spans.get(runId);
            if (!spanData)
                return;
            const { span, operationType, requestModel } = spanData;
            // Extract model name from response (llmOutput) or fall back to request model
            const responseModel = this.extractModelNameFromResponse(output);
            const model = responseModel || requestModel;
            // Update span name if we got a better model name from the response
            if (responseModel && responseModel !== requestModel) {
                span.updateName(`${operationType} ${responseModel}`);
            }
            span.setAttributes({
                [incubating.ATTR_GEN_AI_REQUEST_MODEL]: requestModel || model,
                [incubating.ATTR_GEN_AI_RESPONSE_MODEL]: model,
            });
            // Set response ID if available (providers may expose it in llmOutput)
            const responseId = this.extractResponseId(output);
            if (responseId) {
                span.setAttribute(incubating.ATTR_GEN_AI_RESPONSE_ID, responseId);
            }
            // Collect finish reasons from ALL candidates across ALL generation groups.
            // LLMResult.generations is Generation[][] where outer = prompt batch, inner = n candidates.
            // ATTR_GEN_AI_RESPONSE_FINISH_REASONS should contain one entry per candidate (OTel spec).
            const allFinishReasons = [];
            if (output.generations) {
                for (const group of output.generations) {
                    if (group) {
                        for (const gen of group) {
                            const raw = ((_a = gen === null || gen === void 0 ? void 0 : gen.generationInfo) === null || _a === void 0 ? void 0 : _a.finish_reason) ||
                                ((_b = gen === null || gen === void 0 ? void 0 : gen.generationInfo) === null || _b === void 0 ? void 0 : _b.stop_reason) ||
                                ((_c = gen === null || gen === void 0 ? void 0 : gen.generationInfo) === null || _c === void 0 ? void 0 : _c.done_reason) ||
                                null;
                            if (raw) {
                                allFinishReasons.push((_d = langchainFinishReasonMap[raw]) !== null && _d !== void 0 ? _d : raw);
                            }
                        }
                    }
                }
            }
            // Set finish reasons on span (metadata — NOT gated by traceContent)
            if (allFinishReasons.length > 0) {
                span.setAttribute(incubating.ATTR_GEN_AI_RESPONSE_FINISH_REASONS, allFinishReasons);
            }
            if (this.traceContent &&
                output.generations &&
                output.generations.length > 0) {
                // flatMap over all candidates in all groups — one output message per candidate
                const outputMessages = output.generations.flatMap((group) => (group !== null && group !== void 0 ? group : []).map((gen) => {
                    var _a, _b, _c, _d, _e;
                    const raw = ((_a = gen === null || gen === void 0 ? void 0 : gen.generationInfo) === null || _a === void 0 ? void 0 : _a.finish_reason) ||
                        ((_b = gen === null || gen === void 0 ? void 0 : gen.generationInfo) === null || _b === void 0 ? void 0 : _b.stop_reason) ||
                        ((_c = gen === null || gen === void 0 ? void 0 : gen.generationInfo) === null || _c === void 0 ? void 0 : _c.done_reason) ||
                        null;
                    const genFinishReason = raw
                        ? ((_d = langchainFinishReasonMap[raw]) !== null && _d !== void 0 ? _d : raw)
                        : "";
                    return {
                        role: "assistant",
                        parts: [{ type: "text", content: (_e = gen === null || gen === void 0 ? void 0 : gen.text) !== null && _e !== void 0 ? _e : "" }],
                        finish_reason: genFinishReason,
                    };
                }));
                span.setAttribute(incubating.ATTR_GEN_AI_OUTPUT_MESSAGES, JSON.stringify(outputMessages));
            }
            // Add usage metrics if available
            if ((_e = output.llmOutput) === null || _e === void 0 ? void 0 : _e.usage) {
                const usage = output.llmOutput.usage;
                if (usage.input_tokens || usage.input_tokens === 0) {
                    span.setAttribute(incubating.ATTR_GEN_AI_USAGE_INPUT_TOKENS, usage.input_tokens);
                }
                if (usage.output_tokens || usage.output_tokens === 0) {
                    span.setAttribute(incubating.ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, usage.output_tokens);
                }
                const hasUsage = usage.input_tokens != null || usage.output_tokens != null;
                if (hasUsage) {
                    const totalTokens = (usage.input_tokens || 0) + (usage.output_tokens || 0);
                    span.setAttribute(aiSemanticConventions.SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS, totalTokens);
                }
            }
            // Also check for tokenUsage format (for compatibility)
            if ((_f = output.llmOutput) === null || _f === void 0 ? void 0 : _f.tokenUsage) {
                const usage = output.llmOutput.tokenUsage;
                if (usage.promptTokens || usage.promptTokens === 0) {
                    span.setAttribute(incubating.ATTR_GEN_AI_USAGE_INPUT_TOKENS, usage.promptTokens);
                }
                if (usage.completionTokens || usage.completionTokens === 0) {
                    span.setAttribute(incubating.ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, usage.completionTokens);
                }
                if (usage.totalTokens || usage.totalTokens === 0) {
                    span.setAttribute(aiSemanticConventions.SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS, usage.totalTokens);
                }
            }
            span.setStatus({ code: api.SpanStatusCode.OK });
            span.end();
            this.spans.delete(runId);
        });
    }
    handleLLMError(err, runId, _parentRunId, _tags, _extraParams) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            const spanData = this.spans.get(runId);
            if (!spanData)
                return;
            const { span } = spanData;
            span.recordException(err);
            span.setStatus({ code: api.SpanStatusCode.ERROR, message: err.message });
            span.end();
            this.spans.delete(runId);
        });
    }
    // Parameter order follows @langchain/core >=1.0.0.
    // The 0.3.x signature had a different order (parentRunId at pos 4, runType at pos 7, runName at pos 8).
    // peerDependencies requires >=1.0.0 to ensure this signature matches at runtime.
    handleChainStart(chain, inputs, runId, _runType, _tags, _metadata, runName, _parentRunId, _extra) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            const chainName = ((_a = chain.id) === null || _a === void 0 ? void 0 : _a[chain.id.length - 1]) || "unknown";
            const displayName = runName || chainName;
            // Detect whether this chain is an agent executor vs a regular chain.
            // Both pass runType: undefined in LangChain 1.x, so we use the class name.
            // Only agent executors get invoke_agent; regular chains use "workflow" (custom,
            // no OTel well-known value exists for generic chain execution).
            const isAgent = this.isAgentChain(chainName);
            const operationName = isAgent
                ? incubating.GEN_AI_OPERATION_NAME_VALUE_INVOKE_AGENT
                : "workflow";
            const span = this.tracer.startSpan(`${operationName} ${displayName}`, {
                kind: api.SpanKind.INTERNAL,
            });
            const attributes = {
                [incubating.ATTR_GEN_AI_OPERATION_NAME]: operationName,
                [incubating.ATTR_GEN_AI_PROVIDER_NAME]: "langchain",
                // Backward compatibility
                "traceloop.span.kind": "workflow",
                "traceloop.workflow.name": displayName,
            };
            if (isAgent) {
                attributes[incubating.ATTR_GEN_AI_AGENT_NAME] = displayName;
            }
            span.setAttributes(attributes);
            if (this.traceContent) {
                span.setAttributes({
                    "traceloop.entity.input": JSON.stringify(inputs),
                });
            }
            this.spans.set(runId, {
                span,
                runId,
                operationType: operationName,
            });
        });
    }
    handleChainEnd(outputs, runId, _parentRunId, _tags, _kwargs) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            const spanData = this.spans.get(runId);
            if (!spanData)
                return;
            const { span } = spanData;
            if (this.traceContent) {
                span.setAttributes({
                    "traceloop.entity.output": JSON.stringify(outputs),
                });
            }
            span.setStatus({ code: api.SpanStatusCode.OK });
            span.end();
            this.spans.delete(runId);
        });
    }
    handleChainError(err, runId, _parentRunId, _tags, _kwargs) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            const spanData = this.spans.get(runId);
            if (!spanData)
                return;
            const { span } = spanData;
            span.recordException(err);
            span.setStatus({ code: api.SpanStatusCode.ERROR, message: err.message });
            span.end();
            this.spans.delete(runId);
        });
    }
    handleToolStart(tool, input, runId, _parentRunId, _tags, _metadata, _runName) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            const toolName = ((_a = tool.id) === null || _a === void 0 ? void 0 : _a[tool.id.length - 1]) || "unknown";
            const span = this.tracer.startSpan(`${incubating.GEN_AI_OPERATION_NAME_VALUE_EXECUTE_TOOL} ${toolName}`, {
                kind: api.SpanKind.INTERNAL,
            });
            span.setAttributes({
                [incubating.ATTR_GEN_AI_OPERATION_NAME]: incubating.GEN_AI_OPERATION_NAME_VALUE_EXECUTE_TOOL,
                [incubating.ATTR_GEN_AI_PROVIDER_NAME]: "langchain",
                // Backward compatibility
                "traceloop.span.kind": "task",
                "traceloop.entity.name": toolName,
            });
            if (this.traceContent) {
                span.setAttributes({
                    "traceloop.entity.input": JSON.stringify({ args: [input] }),
                });
            }
            this.spans.set(runId, {
                span,
                runId,
                operationType: incubating.GEN_AI_OPERATION_NAME_VALUE_EXECUTE_TOOL,
            });
        });
    }
    handleToolEnd(output, runId, _parentRunId, _tags) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            const spanData = this.spans.get(runId);
            if (!spanData)
                return;
            const { span } = spanData;
            if (this.traceContent) {
                span.setAttributes({
                    "traceloop.entity.output": JSON.stringify(output),
                });
            }
            span.setStatus({ code: api.SpanStatusCode.OK });
            span.end();
            this.spans.delete(runId);
        });
    }
    handleToolError(err, runId, _parentRunId, _tags) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            const spanData = this.spans.get(runId);
            if (!spanData)
                return;
            const { span } = spanData;
            span.recordException(err);
            span.setStatus({ code: api.SpanStatusCode.ERROR, message: err.message });
            span.end();
            this.spans.delete(runId);
        });
    }
    extractModelFromExtraParams(extraParams) {
        var _a;
        // LangChain passes invocation_params in extraParams for handleChatModelStart/handleLLMStart
        // which contains the model name from the provider SDK (e.g., "gpt-4o-mini")
        const invocationParams = extraParams === null || extraParams === void 0 ? void 0 : extraParams.invocation_params;
        const model = (_a = invocationParams === null || invocationParams === void 0 ? void 0 : invocationParams.model) !== null && _a !== void 0 ? _a : invocationParams === null || invocationParams === void 0 ? void 0 : invocationParams.model_name;
        return model && typeof model === "string" ? model : undefined;
    }
    extractResponseId(output) {
        // Providers may expose response ID in llmOutput (e.g., OpenAI's chatcmpl-xxx)
        if (output.llmOutput) {
            const id = output.llmOutput.id || output.llmOutput.response_id;
            if (id && typeof id === "string") {
                return id;
            }
        }
        return null;
    }
    extractModelNameFromResponse(output) {
        // Follow Python implementation - extract from llm_output first
        if (output.llmOutput) {
            const modelName = output.llmOutput.model_name ||
                output.llmOutput.model_id ||
                output.llmOutput.model;
            if (modelName && typeof modelName === "string") {
                return modelName;
            }
        }
        return null;
    }
    detectVendor(llm) {
        var _a;
        const className = ((_a = llm.id) === null || _a === void 0 ? void 0 : _a[llm.id.length - 1]) || "";
        if (!className) {
            return "langchain";
        }
        // Ordered by specificity (most specific first)
        // Azure (most specific - check first)
        if (["AzureChatOpenAI", "AzureOpenAI", "AzureOpenAIEmbeddings"].includes(className) ||
            className.toLowerCase().includes("azure")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_AZURE_AI_OPENAI;
        }
        // OpenAI
        if (["ChatOpenAI", "OpenAI", "OpenAIEmbeddings"].includes(className) ||
            className.toLowerCase().includes("openai")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_OPENAI;
        }
        // AWS Bedrock
        if (["ChatBedrock", "BedrockEmbeddings", "Bedrock", "BedrockChat"].includes(className) ||
            className.toLowerCase().includes("bedrock") ||
            className.toLowerCase().includes("aws")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_AWS_BEDROCK;
        }
        // Anthropic
        if (["ChatAnthropic", "AnthropicLLM"].includes(className) ||
            className.toLowerCase().includes("anthropic")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_ANTHROPIC;
        }
        // Google Vertex AI (aiplatform.googleapis.com)
        if (["ChatVertexAI", "VertexAI", "VertexAIEmbeddings"].includes(className) ||
            className.toLowerCase().includes("vertex")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_GCP_VERTEX_AI;
        }
        // Google Gemini / AI Studio (generativelanguage.googleapis.com)
        if ([
            "ChatGoogleGenerativeAI",
            "GoogleGenerativeAI",
            "GoogleGenerativeAIEmbeddings",
        ].includes(className) ||
            className.toLowerCase().includes("gemini")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_GCP_GEMINI;
        }
        // Google PaLM / generic Google (fallback)
        if (["GooglePaLM", "ChatGooglePaLM"].includes(className) ||
            className.toLowerCase().includes("google") ||
            className.toLowerCase().includes("palm")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_GCP_GEN_AI;
        }
        // Cohere
        if (["ChatCohere", "CohereEmbeddings", "Cohere"].includes(className) ||
            className.toLowerCase().includes("cohere")) {
            return incubating.GEN_AI_PROVIDER_NAME_VALUE_COHERE;
        }
        // HuggingFace - no OTel constant, use string
        if ([
            "HuggingFacePipeline",
            "HuggingFaceTextGenInference",
            "HuggingFaceEmbeddings",
            "ChatHuggingFace",
        ].includes(className) ||
            className.toLowerCase().includes("huggingface")) {
            return "huggingface";
        }
        // Ollama - no OTel constant, use string
        if (["ChatOllama", "OllamaEmbeddings", "Ollama"].includes(className) ||
            className.toLowerCase().includes("ollama")) {
            return "ollama";
        }
        // Together - no OTel constant, use string
        if (["Together", "ChatTogether"].includes(className) ||
            className.toLowerCase().includes("together")) {
            return "together_ai";
        }
        // Replicate - no OTel constant, use string
        if (["Replicate", "ChatReplicate"].includes(className) ||
            className.toLowerCase().includes("replicate")) {
            return "replicate";
        }
        return "langchain";
    }
    isAgentChain(chainName) {
        const lower = chainName.toLowerCase();
        return (lower.includes("agent") ||
            lower.includes("executor") ||
            lower === "agentexecutor");
    }
    mapMessageTypeToRole(messageType) {
        // Map LangChain message types to standard OpenTelemetry roles
        switch (messageType) {
            case "human":
                return "user";
            case "ai":
                return "assistant";
            case "system":
                return "system";
            case "function":
                return "tool";
            default:
                return messageType;
        }
    }
}

var version = "0.25.0";

/*
 * Copyright Traceloop
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
class LangChainInstrumentation extends instrumentation.InstrumentationBase {
    constructor(config = {}) {
        super("@traceloop/instrumentation-langchain", version, config);
        // Manually instrument CallbackManager immediately since module detection doesn't work
        this.instrumentCallbackManagerDirectly();
    }
    manuallyInstrument({ callbackManagerModule, }) {
        if (callbackManagerModule) {
            this._diag.debug("Manually instrumenting @langchain/core/callbacks/manager");
            this.patchCallbackManager(callbackManagerModule.CallbackManager);
        }
    }
    init() {
        // Return empty array since we handle patching in constructor
        return [];
    }
    instrumentCallbackManagerDirectly() {
        // Patch CJS module
        try {
            const callbackManagerModule = require("@langchain/core/callbacks/manager");
            if (callbackManagerModule === null || callbackManagerModule === void 0 ? void 0 : callbackManagerModule.CallbackManager) {
                this.patchCallbackManager(callbackManagerModule.CallbackManager);
            }
        }
        catch (error) {
            this._diag.debug("Error instrumenting CJS callback manager:", error);
        }
        // Also patch ESM module — in LangChain 1.x, CJS and ESM resolve to
        // different objects, so patching CJS alone doesn't work for ESM consumers.
        import('@langchain/core/callbacks/manager')
            .then((esmModule) => {
            if (esmModule === null || esmModule === void 0 ? void 0 : esmModule.CallbackManager) {
                this.patchCallbackManager(esmModule.CallbackManager);
            }
        })
            .catch((error) => {
            this._diag.debug("Error instrumenting ESM callback manager:", error);
        });
    }
    patchCallbackManager(CallbackManager) {
        const callbackManagerAny = CallbackManager;
        if (callbackManagerAny._configureSync &&
            !callbackManagerAny._traceloopPatched) {
            const originalConfigureSync = callbackManagerAny._configureSync;
            // eslint-disable-next-line @typescript-eslint/no-this-alias
            const self = this;
            callbackManagerAny._configureSync = function (inheritableHandlers, localHandlers, inheritableTags, localTags, inheritableMetadata, localMetadata) {
                // Add our callback handler to inheritable handlers
                const callbackHandler = new TraceloopCallbackHandler(self.tracer, self._shouldSendPrompts());
                const updatedInheritableHandlers = inheritableHandlers && Array.isArray(inheritableHandlers)
                    ? [...inheritableHandlers, callbackHandler]
                    : [callbackHandler];
                return originalConfigureSync.call(this, updatedInheritableHandlers, localHandlers, inheritableTags, localTags, inheritableMetadata, localMetadata);
            };
            // Mark as patched to avoid double patching
            callbackManagerAny._traceloopPatched = true;
        }
    }
    _shouldSendPrompts() {
        const contextShouldSendPrompts = api.context
            .active()
            .getValue(aiSemanticConventions.CONTEXT_KEY_ALLOW_TRACE_CONTENT);
        if (contextShouldSendPrompts !== undefined) {
            return !!contextShouldSendPrompts;
        }
        return this._config.traceContent !== undefined
            ? this._config.traceContent
            : true;
    }
}

exports.LangChainInstrumentation = LangChainInstrumentation;
exports.TraceloopCallbackHandler = TraceloopCallbackHandler;
