// @traceloop/instrumentation-cohere/dist/index.mjs
import { __awaiter, __asyncValues } from 'tslib';
import { trace, context, SpanStatusCode, SpanKind } from '@opentelemetry/api';
import { InstrumentationBase, InstrumentationNodeModuleDefinition, safeExecuteInTheMiddle } from '@opentelemetry/instrumentation';
import { SpanAttributes, LLMRequestTypeValues, CONTEXT_KEY_ALLOW_TRACE_CONTENT } from '@traceloop/ai-semantic-conventions';
import { ATTR_GEN_AI_REQUEST_MODEL, ATTR_GEN_AI_REQUEST_TOP_P, ATTR_GEN_AI_REQUEST_TEMPERATURE, ATTR_GEN_AI_REQUEST_MAX_TOKENS, ATTR_GEN_AI_PROMPT, ATTR_GEN_AI_COMPLETION, ATTR_GEN_AI_USAGE_PROMPT_TOKENS, ATTR_GEN_AI_USAGE_COMPLETION_TOKENS, ATTR_GEN_AI_SYSTEM } from '@opentelemetry/semantic-conventions/incubating';

var version = "0.26.0";

class CohereInstrumentation extends InstrumentationBase {
    constructor(config = {}) {
        super("@traceloop/instrumentation-cohere", version, config);
    }
    setConfig(config = {}) {
        super.setConfig(config);
    }
    init() {
        const module = new InstrumentationNodeModuleDefinition("cohere-ai", [">=7.7.5"], this.wrap.bind(this), this.unwrap.bind(this));
        return module;
    }
    manuallyInstrument(module) {
        this._diag.debug(`Manually patching cohere-ai`);
        this.wrap(module);
    }
    wrap(module, moduleVersion) {
        this._diag.debug(`Patching cohere-ai@${moduleVersion}`);
        this._wrap(module.CohereClient.prototype, "generate", this.wrapperMethod("completion", false));
        this._wrap(module.CohereClient.prototype, "generateStream", this.wrapperMethod("completion", true));
        this._wrap(module.CohereClient.prototype, "chat", this.wrapperMethod("chat", false));
        this._wrap(module.CohereClient.prototype, "chatStream", this.wrapperMethod("chat", true));
        this._wrap(module.CohereClient.prototype, "rerank", this.wrapperMethod("rerank", false));
        return module;
    }
    unwrap(module, moduleVersion) {
        this._diag.debug(`Unpatching @cohere-ai@${moduleVersion}`);
        this._unwrap(module.CohereClient.prototype, "generateStream");
        this._unwrap(module.CohereClient.prototype, "chat");
        this._unwrap(module.CohereClient.prototype, "chatStream");
        this._unwrap(module.CohereClient.prototype, "rerank");
    }
    wrapperMethod(type, streaming) {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const plugin = this;
        // eslint-disable-next-line
        return (original) => {
            return function method(...args) {
                const span = plugin._startSpan({
                    params: args[0],
                    type,
                });
                const execContext = trace.setSpan(context.active(), span);
                const execPromise = safeExecuteInTheMiddle(() => {
                    return context.with(execContext, () => {
                        return original.apply(this, args);
                    });
                }, (e) => {
                    if (e) {
                        plugin._diag.error("Error in cohere instrumentation", e);
                    }
                });
                const wrappedPromise = plugin._wrapPromise(type, streaming, span, execPromise);
                return context.bind(execContext, wrappedPromise);
            };
        };
    }
    _wrapPromise(type, streaming, span, promise) {
        return promise
            .then((result) => __awaiter(this, void 0, void 0, function* () {
            const awaitedResult = yield result;
            if (type === "completion" && streaming) {
                yield this._endSpan({
                    type,
                    span,
                    streaming,
                    result: (yield awaitedResult),
                });
            }
            else if (type === "completion" && !streaming) {
                yield this._endSpan({
                    type,
                    span,
                    streaming,
                    result: awaitedResult,
                });
            }
            else if (type === "chat" && streaming) {
                yield this._endSpan({
                    type,
                    span,
                    streaming,
                    result: (yield awaitedResult),
                });
            }
            else if (type === "chat" && !streaming) {
                yield this._endSpan({
                    type,
                    span,
                    streaming,
                    result: awaitedResult,
                });
            }
            else if (type === "rerank" && !streaming) {
                yield this._endSpan({
                    type,
                    span,
                    streaming,
                    result: awaitedResult,
                });
            }
            return new Promise((resolve) => resolve(result));
        }))
            .catch((error) => {
            return new Promise((_, reject) => {
                span.setStatus({
                    code: SpanStatusCode.ERROR,
                    message: error.message,
                });
                span.recordException(error);
                span.end();
                reject(error);
            });
        });
    }
    _startSpan({ params, type, }) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const attributes = {
            [ATTR_GEN_AI_SYSTEM]: "Cohere",
            [SpanAttributes.LLM_REQUEST_TYPE]: this._getLlmRequestTypeByMethod(type),
        };
        try {
            const model = (_a = params.model) !== null && _a !== void 0 ? _a : "command";
            attributes[ATTR_GEN_AI_REQUEST_MODEL] = model;
            if (!("query" in params)) {
                attributes[ATTR_GEN_AI_REQUEST_TOP_P] = params.p;
                attributes[SpanAttributes.LLM_TOP_K] = params.k;
                attributes[ATTR_GEN_AI_REQUEST_TEMPERATURE] = params.temperature;
                attributes[SpanAttributes.LLM_FREQUENCY_PENALTY] =
                    params.frequencyPenalty;
                attributes[SpanAttributes.LLM_PRESENCE_PENALTY] =
                    params.presencePenalty;
                attributes[ATTR_GEN_AI_REQUEST_MAX_TOKENS] = params.maxTokens;
            }
            else {
                attributes["topN"] = params["topN"];
                attributes["maxChunksPerDoc"] = params["maxChunksPerDoc"];
            }
            if (this._shouldSendPrompts()) {
                if (type === "completion" && "prompt" in params) {
                    attributes[`${ATTR_GEN_AI_PROMPT}.0.role`] = "user";
                    attributes[`${ATTR_GEN_AI_PROMPT}.0.user`] = params.prompt;
                }
                else if (type === "chat" && "message" in params) {
                    (_b = params.chatHistory) === null || _b === void 0 ? void 0 : _b.forEach((msg, index) => {
                        attributes[`${ATTR_GEN_AI_PROMPT}.${index}.role`] = msg.role;
                        if (msg.role !== "TOOL") {
                            attributes[`${ATTR_GEN_AI_PROMPT}.${index}.content`] =
                                msg.message;
                        }
                    });
                    attributes[`${ATTR_GEN_AI_PROMPT}.${(_d = (_c = params.chatHistory) === null || _c === void 0 ? void 0 : _c.length) !== null && _d !== void 0 ? _d : 0}.role`] = "user";
                    attributes[`${ATTR_GEN_AI_PROMPT}.${(_f = (_e = params.chatHistory) === null || _e === void 0 ? void 0 : _e.length) !== null && _f !== void 0 ? _f : 0}.user`] = params.message;
                }
                else if (type === "rerank" && "query" in params) {
                    attributes[`${ATTR_GEN_AI_PROMPT}.0.role`] = "user";
                    attributes[`${ATTR_GEN_AI_PROMPT}.0.user`] = params.query;
                    params.documents.forEach((doc, index) => {
                        attributes[`documents.${index}.index`] =
                            typeof doc === "string" ? doc : doc.text;
                    });
                }
            }
        }
        catch (e) {
            this._diag.debug(e);
            (_h = (_g = this._config).exceptionLogger) === null || _h === void 0 ? void 0 : _h.call(_g, e);
        }
        return this.tracer.startSpan(`cohere.${type}`, {
            kind: SpanKind.CLIENT,
            attributes,
        });
    }
    _endSpan(_a) {
        return __awaiter(this, arguments, void 0, function* ({ type, span, streaming, result, }) {
            var _b, result_1, result_1_1, _c, result_2, result_2_1;
            var _d, e_1, _e, _f, _g, e_2, _h, _j;
            if (type === "completion") {
                if (streaming) {
                    try {
                        for (_b = true, result_1 = __asyncValues(result); result_1_1 = yield result_1.next(), _d = result_1_1.done, !_d; _b = true) {
                            _f = result_1_1.value;
                            _b = false;
                            const message = _f;
                            if (message.eventType === "stream-end") {
                                this._setResponseSpanForGenerate(span, message.response);
                            }
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (!_b && !_d && (_e = result_1.return)) yield _e.call(result_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
                else if ("generations" in result) {
                    this._setResponseSpanForGenerate(span, result);
                }
            }
            else if (type === "chat") {
                if (streaming) {
                    try {
                        for (_c = true, result_2 = __asyncValues(result); result_2_1 = yield result_2.next(), _g = result_2_1.done, !_g; _c = true) {
                            _j = result_2_1.value;
                            _c = false;
                            const message = _j;
                            if (message.eventType === "stream-end") {
                                this._setResponseSpanForChat(span, message.response);
                            }
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (!_c && !_g && (_h = result_2.return)) yield _h.call(result_2);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                }
                else if ("text" in result) {
                    this._setResponseSpanForChat(span, result);
                }
            }
            else if (type === "rerank") {
                if ("results" in result) {
                    this._setResponseSpanForRerank(span, result);
                }
            }
            span.setStatus({ code: SpanStatusCode.OK });
            span.end();
        });
    }
    _setResponseSpanForRerank(span, result) {
        var _a, _b, _c, _d, _e, _f;
        try {
            if ("meta" in result) {
                if (((_b = (_a = result.meta) === null || _a === void 0 ? void 0 : _a.billedUnits) === null || _b === void 0 ? void 0 : _b.searchUnits) !== undefined) {
                    span.setAttribute(SpanAttributes.LLM_USAGE_TOTAL_TOKENS, (_d = (_c = result.meta) === null || _c === void 0 ? void 0 : _c.billedUnits) === null || _d === void 0 ? void 0 : _d.searchUnits);
                }
                if (this._shouldSendPrompts()) {
                    result.results.forEach((each, idx) => {
                        var _a;
                        span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.${idx}.relevanceScore`, each.relevanceScore);
                        if (each.document && ((_a = each.document) === null || _a === void 0 ? void 0 : _a.text)) {
                            span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.${idx}.content`, each.document.text);
                        }
                    });
                }
                else {
                    result.results.forEach((each, idx) => {
                        span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.${idx}.content`, each.index);
                        span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.${idx}.relevanceScore`, each.relevanceScore);
                    });
                }
            }
        }
        catch (e) {
            this._diag.debug(e);
            (_f = (_e = this._config).exceptionLogger) === null || _f === void 0 ? void 0 : _f.call(_e, e);
        }
    }
    _setResponseSpanForChat(span, result) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        try {
            if ("token_count" in result && typeof result.token_count === "object") {
                if (result.token_count &&
                    "prompt_tokens" in result.token_count &&
                    typeof result.token_count.prompt_tokens === "number") {
                    span.setAttribute(ATTR_GEN_AI_USAGE_PROMPT_TOKENS, (_a = result.token_count) === null || _a === void 0 ? void 0 : _a.prompt_tokens);
                }
                if (result.token_count &&
                    "response_tokens" in result.token_count &&
                    typeof result.token_count.response_tokens === "number") {
                    span.setAttribute(ATTR_GEN_AI_USAGE_COMPLETION_TOKENS, (_b = result.token_count) === null || _b === void 0 ? void 0 : _b.response_tokens);
                }
                if (result.token_count &&
                    "total_tokens" in result.token_count &&
                    typeof result.token_count.total_tokens === "number") {
                    span.setAttribute(SpanAttributes.LLM_USAGE_TOTAL_TOKENS, (_c = result.token_count) === null || _c === void 0 ? void 0 : _c.total_tokens);
                }
            }
            if (this._shouldSendPrompts()) {
                span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.role`, "assistant");
                span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.content`, result.text);
                if ((_d = result.searchQueries) === null || _d === void 0 ? void 0 : _d[0].text) {
                    span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.searchQuery`, (_e = result.searchQueries) === null || _e === void 0 ? void 0 : _e[0].text);
                }
                if ((_f = result.searchResults) === null || _f === void 0 ? void 0 : _f.length) {
                    result.searchResults.forEach((searchResult, index) => {
                        if (searchResult.searchQuery) {
                            span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.searchResult.${index}.text`, searchResult.searchQuery.text);
                        }
                        span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.searchResult.${index}.connector`, searchResult.connector.id);
                    });
                }
            }
            if ("finishReason" in result && typeof result.finishReason === "string") {
                span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.finish_reason`, result.finishReason);
            }
        }
        catch (e) {
            this._diag.debug(e);
            (_h = (_g = this._config).exceptionLogger) === null || _h === void 0 ? void 0 : _h.call(_g, e);
        }
    }
    _setResponseSpanForGenerate(span, result) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
        try {
            if (result && "meta" in result) {
                if (typeof ((_b = (_a = result.meta) === null || _a === void 0 ? void 0 : _a.billedUnits) === null || _b === void 0 ? void 0 : _b.inputTokens) === "number") {
                    span.setAttribute(ATTR_GEN_AI_USAGE_PROMPT_TOKENS, (_d = (_c = result.meta) === null || _c === void 0 ? void 0 : _c.billedUnits) === null || _d === void 0 ? void 0 : _d.inputTokens);
                }
                if (typeof ((_f = (_e = result.meta) === null || _e === void 0 ? void 0 : _e.billedUnits) === null || _f === void 0 ? void 0 : _f.outputTokens) === "number") {
                    span.setAttribute(ATTR_GEN_AI_USAGE_COMPLETION_TOKENS, (_h = (_g = result.meta) === null || _g === void 0 ? void 0 : _g.billedUnits) === null || _h === void 0 ? void 0 : _h.outputTokens);
                }
                if (typeof ((_k = (_j = result.meta) === null || _j === void 0 ? void 0 : _j.billedUnits) === null || _k === void 0 ? void 0 : _k.inputTokens) === "number" &&
                    typeof ((_m = (_l = result.meta) === null || _l === void 0 ? void 0 : _l.billedUnits) === null || _m === void 0 ? void 0 : _m.outputTokens) === "number") {
                    span.setAttribute(SpanAttributes.LLM_USAGE_TOTAL_TOKENS, ((_p = (_o = result.meta) === null || _o === void 0 ? void 0 : _o.billedUnits) === null || _p === void 0 ? void 0 : _p.inputTokens) +
                        ((_r = (_q = result.meta) === null || _q === void 0 ? void 0 : _q.billedUnits) === null || _r === void 0 ? void 0 : _r.outputTokens));
                }
            }
            if (this._shouldSendPrompts() && result.generations) {
                span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.role`, "assistant");
                span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.content`, result.generations[0].text);
            }
            if (result.generations &&
                "finish_reason" in result.generations[0] &&
                typeof result.generations[0].finish_reason === "string") {
                span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.finish_reason`, result.generations[0].finish_reason);
            }
            if (result.generations &&
                "finishReason" in result.generations[0] &&
                typeof result.generations[0].finishReason === "string") {
                span.setAttribute(`${ATTR_GEN_AI_COMPLETION}.0.finish_reason`, result.generations[0].finishReason);
            }
        }
        catch (e) {
            this._diag.debug(e);
            (_t = (_s = this._config).exceptionLogger) === null || _t === void 0 ? void 0 : _t.call(_s, e);
        }
    }
    _getLlmRequestTypeByMethod(type) {
        if (type === "chat")
            return LLMRequestTypeValues.CHAT;
        else if (type === "completion")
            return LLMRequestTypeValues.COMPLETION;
        else if (type === "rerank")
            return LLMRequestTypeValues.RERANK;
        else
            return LLMRequestTypeValues.UNKNOWN;
    }
    _shouldSendPrompts() {
        const contextShouldSendPrompts = context
            .active()
            .getValue(CONTEXT_KEY_ALLOW_TRACE_CONTENT);
        if (contextShouldSendPrompts !== undefined) {
            return contextShouldSendPrompts;
        }
        return this._config.traceContent !== undefined
            ? this._config.traceContent
            : true;
    }
}

export { CohereInstrumentation };
