// @latitude-data/telemetry/dist/index.js
import { t as getLatitudeTracer } from "./tracer-C3coEEca.js";
import { ProxyTracerProvider, context, createContextKey, propagation, trace } from "@opentelemetry/api";
import { AsyncLocalStorageContextManager } from "@opentelemetry/context-async-hooks";
import { CompositePropagator, W3CBaggagePropagator, W3CTraceContextPropagator } from "@opentelemetry/core";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { BatchSpanProcessor, NodeTracerProvider, SimpleSpanProcessor } from "@opentelemetry/sdk-trace-node";
import { ATTR_SERVICE_NAME } from "@opentelemetry/semantic-conventions";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
//#region src/sdk/_deprecation.ts
/**
* Internal helpers for one-time deprecation warnings.
*
* Lives in its own module so both `init.ts` and `context.ts` can depend on it
* without forming a circular import (`init` already imports the span processor,
* which imports `context`). Mirrors the Python `_deprecation.py` module.
*/
let projectSlugDeprecationWarned = false;
function warnProjectSlugDeprecated(site) {
	if (projectSlugDeprecationWarned) return;
	projectSlugDeprecationWarned = true;
	console.warn(`[Latitude] ${site === "constructor" ? "`projectSlug`" : "`projectSlug` on capture()"} is deprecated and will be removed in a future release — rename it to \`project\`. Both work for now; when both are passed, \`project\` wins.`);
}
//#endregion
//#region src/sdk/context.ts
const LATITUDE_CONTEXT_KEY = createContextKey("latitude-internal-context");
const CAPTURE_TRACER_NAME = "so.latitude.instrumentation.capture";
function recordSpanExceptionForDatadog(span, error) {
	const err = error instanceof Error ? error : new Error(String(error));
	span.recordException(err);
	span.setAttributes({
		"error.message": err.message,
		"error.stack": err.stack ?? "",
		"error.type": err.constructor.name
	});
}
function getLatitudeContext(ctx) {
	return ctx.getValue(LATITUDE_CONTEXT_KEY);
}
function mergeArrays(a, b) {
	if (!a && !b) return void 0;
	if (!a) return b;
	if (!b) return a;
	return [...new Set([...a, ...b])];
}
function shouldReuseActiveLatitudeTrace(currentContext) {
	return getLatitudeContext(currentContext) !== void 0;
}
function capture(name, fn, options = {}) {
	const currentContext = context.active();
	const existingData = getLatitudeContext(currentContext);
	const shouldReuseTrace = shouldReuseActiveLatitudeTrace(currentContext);
	const parentContext = shouldReuseTrace ? currentContext : trace.deleteSpan(currentContext);
	if (options.project === void 0 && options.projectSlug !== void 0) warnProjectSlugDeprecated("capture");
	const projectFromOptions = options.project ?? options.projectSlug;
	const mergedData = {
		name: options.name ?? name,
		tags: mergeArrays(existingData?.tags, options.tags),
		metadata: {
			...existingData?.metadata,
			...options.metadata
		},
		sessionId: options.sessionId ?? existingData?.sessionId,
		userId: options.userId ?? existingData?.userId,
		userEmail: options.userEmail ?? existingData?.userEmail,
		project: projectFromOptions ?? existingData?.project
	};
	const newContext = parentContext.setValue(LATITUDE_CONTEXT_KEY, mergedData);
	if (trace.getSpan(currentContext) && shouldReuseTrace) return context.with(newContext, fn);
	return trace.getTracer(CAPTURE_TRACER_NAME).startActiveSpan(name, { attributes: { "latitude.capture.root": true } }, newContext, (span) => {
		let result;
		try {
			result = fn();
		} catch (error) {
			recordSpanExceptionForDatadog(span, error);
			span.end();
			throw error;
		}
		if (result instanceof Promise) return result.catch((error) => {
			recordSpanExceptionForDatadog(span, error);
			throw error;
		}).finally(() => {
			span.end();
		});
		span.end();
		return result;
	});
}
//#endregion
//#region src/sdk/instrumentations.ts
const DOCS_URL = "https://github.com/latitude-dev/latitude-llm/tree/main/packages/telemetry/typescript#readme";
/** Exported for direct unit testing — internal otherwise. */
const normalizeOpenAI = (mod) => {
	const ns = mod;
	return ns?.OpenAI ?? ns?.default ?? mod;
};
/** Exported for direct unit testing — internal otherwise. */
const normalizeAnthropic = (mod) => {
	return mod !== null && typeof mod === "object" && "Anthropic" in mod ? mod : { Anthropic: mod };
};
const INTEGRATIONS = {
	openai: {
		loadCtor: async () => (await import("./instrumentation-DFz-ac4w.js")).OpenAIInstrumentationWithResponses,
		packageName: "@traceloop/instrumentation-openai",
		normalize: normalizeOpenAI,
		enrichTokens: true
	},
	"openai-agents": {
		loadCtor: async () => (await import("./instrumentation-S6F3OJq3.js")).OpenAIAgentsInstrumentation,
		packageName: "@openai/agents"
	},
	anthropic: {
		loadCtor: async () => (await import("@traceloop/instrumentation-anthropic")).AnthropicInstrumentation,
		packageName: "@traceloop/instrumentation-anthropic",
		normalize: normalizeAnthropic
	},
	bedrock: {
		loadCtor: async () => (await import("@traceloop/instrumentation-bedrock")).BedrockInstrumentation,
		packageName: "@traceloop/instrumentation-bedrock"
	},
	cohere: {
		loadCtor: async () => (await import("@traceloop/instrumentation-cohere")).CohereInstrumentation,
		packageName: "@traceloop/instrumentation-cohere"
	},
	langchain: {
		loadCtor: async () => (await import("@traceloop/instrumentation-langchain")).LangChainInstrumentation,
		packageName: "@traceloop/instrumentation-langchain"
	},
	llamaindex: {
		loadCtor: async () => (await import("@traceloop/instrumentation-llamaindex")).LlamaIndexInstrumentation,
		packageName: "@traceloop/instrumentation-llamaindex"
	},
	togetherai: {
		loadCtor: async () => (await import("@traceloop/instrumentation-together")).TogetherInstrumentation,
		packageName: "@traceloop/instrumentation-together",
		enrichTokens: false
	},
	vertexai: {
		loadCtor: async () => (await import("@traceloop/instrumentation-vertexai")).VertexAIInstrumentation,
		packageName: "@traceloop/instrumentation-vertexai"
	},
	aiplatform: {
		loadCtor: async () => (await import("@traceloop/instrumentation-vertexai")).AIPlatformInstrumentation,
		packageName: "@traceloop/instrumentation-vertexai"
	}
};
function isPlainObject(value) {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}
async function createLatitudeInstrumentations(input) {
	if (!isPlainObject(input)) throw new TypeError(`[Latitude] instrumentations must be an object mapping integration names to LLM SDK modules (e.g. { openai: OpenAI, anthropic: AnthropicSDK }). Received: ${JSON.stringify(input)}. See ${DOCS_URL}.`);
	const result = [];
	for (const [name, mod] of Object.entries(input)) {
		if (mod === void 0) continue;
		const def = INTEGRATIONS[name];
		if (!def) throw new TypeError(`[Latitude] instrumentations: unknown integration "${name}". Expected one of: ${Object.keys(INTEGRATIONS).join(", ")}. See ${DOCS_URL}.`);
		let Ctor;
		try {
			Ctor = await def.loadCtor();
		} catch {
			console.warn(`[Latitude] Instrumentation package not installed for ${name}: ${def.packageName}. Add it as a dependency to enable this instrumentation.`);
			continue;
		}
		const inst = new Ctor(def.enrichTokens !== void 0 ? { enrichTokens: def.enrichTokens } : void 0);
		const target = def.normalize ? def.normalize(mod) : mod;
		inst.manuallyInstrument?.(target);
		result.push(inst);
	}
	return result;
}
/**
* Registers LLM instrumentations with the global OpenTelemetry instrumentation registry.
*
* @example
* ```typescript
* import OpenAI from "openai"
* import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node"
* import { LatitudeSpanProcessor, registerLatitudeInstrumentations } from "@latitude-data/telemetry"
*
* const provider = new NodeTracerProvider({
*   spanProcessors: [new LatitudeSpanProcessor(apiKey, project)],
* })
*
* await registerLatitudeInstrumentations({
*   instrumentations: { openai: OpenAI },
*   tracerProvider: provider,
* })
*
* provider.register()
* ```
*/
async function registerLatitudeInstrumentations(options) {
	registerInstrumentations({
		instrumentations: await createLatitudeInstrumentations(options.instrumentations),
		tracerProvider: options.tracerProvider
	});
}
//#endregion
//#region src/constants/attributes.ts
const ATTRIBUTES = {
	name: "latitude.capture.name",
	tags: "latitude.tags",
	metadata: "latitude.metadata",
	sessionId: "session.id",
	userId: "user.id",
	userEmail: "user.email",
	project: "latitude.project"
};
//#endregion
//#region src/env/env.ts
const PRODUCTION_EXPORTER_URL = "https://ingest.latitude.so";
function getExporterUrl() {
	return process.env.LATITUDE_TELEMETRY_URL ?? PRODUCTION_EXPORTER_URL;
}
const env = { EXPORTER_URL: getExporterUrl() };
//#endregion
//#region src/sdk/redact.ts
var RedactSpanProcessor = class {
	options;
	constructor(options) {
		this.options = options;
		if (!options.mask) this.options.mask = (_attribute, _value) => "******";
	}
	onStart(_span, _context) {}
	onEnd(span) {
		Object.assign(span.attributes, this.redactAttributes(span.attributes));
		for (const event of span.events) {
			if (!event.attributes) continue;
			Object.assign(event.attributes, this.redactAttributes(event.attributes));
		}
		for (const link of span.links) {
			if (!link.attributes) continue;
			Object.assign(link.attributes, this.redactAttributes(link.attributes));
		}
	}
	forceFlush() {
		return Promise.resolve();
	}
	shutdown() {
		return Promise.resolve();
	}
	shouldRedact(attribute) {
		return this.options.attributes.some((pattern) => {
			if (typeof pattern === "string") return attribute === pattern;
			else if (pattern instanceof RegExp) return pattern.test(attribute);
			return false;
		});
	}
	redactAttributes(attributes) {
		const redacted = {};
		for (const [key, value] of Object.entries(attributes)) if (this.shouldRedact(key)) redacted[key] = this.options.mask?.(key, value);
		return redacted;
	}
};
const DEFAULT_REDACT_SPAN_PROCESSOR = () => new RedactSpanProcessor({ attributes: [
	/^http\.request\.header\.authorization$/i,
	/^http\.request\.header\.cookie$/i,
	/^http\.request\.header\.x[-_]api[-_]key$/i,
	/^db\.statement$/i
] });
//#endregion
//#region src/sdk/span-filter.ts
/** OpenTelemetry GenAI semantic convention attribute prefix. */
const GEN_AI_PREFIX = "gen_ai.";
/** Legacy / OpenInference-style LLM attribute prefix. */
const LLM_PREFIX = "llm.";
const OPENINFERENCE_KIND = "openinference.span.kind";
/** OTel Python instrumentation scope prefixes for LLM-related instrumentors we support. */
const OTEL_LLM_INSTRUMENTATION_SCOPE_PREFIXES = [
	"opentelemetry.instrumentation.alephalpha",
	"opentelemetry.instrumentation.anthropic",
	"opentelemetry.instrumentation.bedrock",
	"opentelemetry.instrumentation.cohere",
	"opentelemetry.instrumentation.crewai",
	"opentelemetry.instrumentation.google_generativeai",
	"opentelemetry.instrumentation.groq",
	"opentelemetry.instrumentation.haystack",
	"opentelemetry.instrumentation.langchain",
	"opentelemetry.instrumentation.llamaindex",
	"opentelemetry.instrumentation.mistralai",
	"opentelemetry.instrumentation.ollama",
	"opentelemetry.instrumentation.openai",
	"opentelemetry.instrumentation.replicate",
	"opentelemetry.instrumentation.sagemaker",
	"opentelemetry.instrumentation.together",
	"opentelemetry.instrumentation.transformers",
	"opentelemetry.instrumentation.vertexai",
	"opentelemetry.instrumentation.watsonx",
	"openinference.instrumentation"
];
/** Substrings in scope names that indicate LLM / GenAI instrumentation (e.g. Traceloop JS). */
const LLM_SCOPE_SUBSTRINGS = [
	"openinference",
	"traceloop",
	"langsmith",
	"litellm"
];
/**
* Builds the export predicate from loose option fields (`exactOptionalPropertyTypes`-safe call sites).
*/
function buildShouldExportSpanFromFields(fields) {
	return buildShouldExportSpan({
		...fields.disableSmartFilter !== void 0 ? { disableSmartFilter: fields.disableSmartFilter } : {},
		...fields.shouldExportSpan !== void 0 ? { shouldExportSpan: fields.shouldExportSpan } : {},
		...fields.blockedInstrumentationScopes !== void 0 ? { blockedInstrumentationScopes: fields.blockedInstrumentationScopes } : {}
	});
}
function attributeKeys(span) {
	const attrs = span.attributes;
	if (!attrs || typeof attrs !== "object") return [];
	return Object.keys(attrs);
}
function instrumentationScopeName(span) {
	return span.instrumentationScope?.name ?? "";
}
/** True if the span uses OpenTelemetry GenAI semantic conventions or common LLM attribute namespaces. */
function isGenAiOrLlmAttributeSpan(span) {
	for (const key of attributeKeys(span)) {
		if (key.startsWith(GEN_AI_PREFIX) || key.startsWith(LLM_PREFIX)) return true;
		if (key === OPENINFERENCE_KIND || key.startsWith("openinference.")) return true;
		if (key.startsWith("ai.")) return true;
		if (key.startsWith("eve.") || key.startsWith("flue.")) return true;
		if (key.startsWith("latitude.")) return true;
	}
	return false;
}
/** True if the span was created with Latitude's tracer scopes. */
function isLatitudeInstrumentationSpan(span) {
	const name = instrumentationScopeName(span);
	return name === "so.latitude.instrumentation" || name.startsWith(`so.latitude.instrumentation.`);
}
function isKnownLlmInstrumentationScope(span) {
	const name = instrumentationScopeName(span);
	if (!name) return false;
	for (const prefix of OTEL_LLM_INSTRUMENTATION_SCOPE_PREFIXES) if (name === prefix || name.startsWith(`${prefix}.`)) return true;
	const lower = name.toLowerCase();
	for (const part of LLM_SCOPE_SUBSTRINGS) if (lower.includes(part)) return true;
	return false;
}
/**
* Default export predicate (smart filter): Latitude scopes, GenAI / LLM attributes,
* or known LLM instrumentation scopes.
*/
function isDefaultExportSpan(span) {
	if (isLatitudeInstrumentationSpan(span)) return true;
	if (isGenAiOrLlmAttributeSpan(span)) return true;
	if (isKnownLlmInstrumentationScope(span)) return true;
	return false;
}
function buildShouldExportSpan(options) {
	if (options.disableSmartFilter) return () => true;
	const blocked = new Set(options.blockedInstrumentationScopes ?? []);
	const extra = options.shouldExportSpan;
	return (span) => {
		const scope = instrumentationScopeName(span);
		if (blocked.has(scope)) return false;
		if (isDefaultExportSpan(span)) return true;
		if (extra?.(span)) return true;
		return false;
	};
}
/**
* Drops spans that fail the export predicate before passing them to the inner processor.
* Inner processor should perform redaction and export.
*/
var ExportFilterSpanProcessor = class {
	shouldExport;
	inner;
	constructor(shouldExport, inner) {
		this.shouldExport = shouldExport;
		this.inner = inner;
	}
	onStart(span, parentContext) {
		this.inner.onStart(span, parentContext);
	}
	onEnd(span) {
		if (!this.shouldExport(span)) return;
		this.inner.onEnd(span);
	}
	forceFlush() {
		return this.inner.forceFlush();
	}
	shutdown() {
		return this.inner.shutdown();
	}
};
/** Runs optional redaction then the export processor (batch/simple). */
var RedactThenExportSpanProcessor = class {
	redact;
	exportProcessor;
	constructor(redact, exportProcessor) {
		this.redact = redact;
		this.exportProcessor = exportProcessor;
	}
	onStart(span, parentContext) {
		this.redact?.onStart(span, parentContext);
		this.exportProcessor.onStart(span, parentContext);
	}
	onEnd(span) {
		this.redact?.onEnd(span);
		this.exportProcessor.onEnd(span);
	}
	forceFlush() {
		return this.exportProcessor.forceFlush();
	}
	shutdown() {
		return this.exportProcessor.shutdown();
	}
};
//#endregion
//#region src/sdk/processor.ts
var ServiceNameResourceExporter = class {
	overlay;
	constructor(inner, serviceName) {
		this.inner = inner;
		this.overlay = resourceFromAttributes({ [ATTR_SERVICE_NAME]: serviceName });
	}
	export(spans, resultCallback) {
		const overlay = this.overlay;
		const overridden = spans.map((span) => new Proxy(span, { get(target, prop, receiver) {
			if (prop === "resource") return target.resource.merge(overlay);
			return Reflect.get(target, prop, receiver);
		} }));
		this.inner.export(overridden, resultCallback);
	}
	shutdown() {
		return this.inner.shutdown();
	}
	forceFlush() {
		return this.inner.forceFlush?.() ?? Promise.resolve();
	}
};
var LatitudeSpanProcessor = class {
	tail;
	constructor(apiKey, project, options) {
		if (!apiKey || apiKey.trim() === "") throw new Error("[Latitude] apiKey is required and cannot be empty");
		const normalizedProject = project && project.trim() !== "" ? project : void 0;
		const headers = {
			Authorization: `Bearer ${apiKey}`,
			"Content-Type": "application/json"
		};
		if (normalizedProject) headers["X-Latitude-Project"] = normalizedProject;
		const baseExporter = options?.exporter ?? new OTLPTraceExporter({
			url: `${env.EXPORTER_URL}/v1/traces`,
			headers,
			timeoutMillis: 3e4
		});
		const rawServiceName = options?.serviceName?.trim();
		const exporter = rawServiceName ? new ServiceNameResourceExporter(baseExporter, rawServiceName) : baseExporter;
		const redact = options?.disableRedact ? null : options?.redact ? new RedactSpanProcessor(options.redact) : DEFAULT_REDACT_SPAN_PROCESSOR();
		const batchOrSimple = options?.disableBatch ? new SimpleSpanProcessor(exporter) : new BatchSpanProcessor(exporter);
		const shouldExport = buildShouldExportSpanFromFields({
			disableSmartFilter: options?.disableSmartFilter,
			shouldExportSpan: options?.shouldExportSpan,
			blockedInstrumentationScopes: options?.blockedInstrumentationScopes
		});
		const redactThenExport = new RedactThenExportSpanProcessor(redact, batchOrSimple);
		this.tail = new ExportFilterSpanProcessor(shouldExport, redactThenExport);
	}
	onStart(span, parentContext) {
		const latitudeData = getLatitudeContext(parentContext);
		if (latitudeData) {
			if (latitudeData.name) {
				span.setAttribute(ATTRIBUTES.name, latitudeData.name);
				if (span.attributes["latitude.capture.root"]) span.updateName(latitudeData.name);
			}
			if (latitudeData.tags && latitudeData.tags.length > 0) span.setAttribute(ATTRIBUTES.tags, JSON.stringify(latitudeData.tags));
			if (latitudeData.metadata && Object.keys(latitudeData.metadata).length > 0) span.setAttribute(ATTRIBUTES.metadata, JSON.stringify(latitudeData.metadata));
			if (latitudeData.sessionId) span.setAttribute(ATTRIBUTES.sessionId, latitudeData.sessionId);
			if (latitudeData.userId) span.setAttribute(ATTRIBUTES.userId, latitudeData.userId);
			if (latitudeData.userEmail) span.setAttribute(ATTRIBUTES.userEmail, latitudeData.userEmail);
			if (latitudeData.project) span.setAttribute(ATTRIBUTES.project, latitudeData.project);
		}
		this.tail.onStart(span, parentContext);
	}
	onEnd(span) {
		this.tail.onEnd(span);
	}
	async forceFlush() {
		await this.tail.forceFlush();
	}
	async shutdown() {
		await this.tail.shutdown();
	}
};
//#endregion
//#region src/sdk/init.ts
const SERVICE_NAME = process.env.npm_package_name || "unknown";
const DETECT_PROBE = "@latitude-data/telemetry-detect";
/** Module-level flag to prevent duplicate signal handler registration on repeated Latitude construction */
let shutdownHandlersRegistered = false;
function getRegisteredTracerProvider() {
	const provider = trace.getTracerProvider();
	if (provider instanceof ProxyTracerProvider) return provider.getDelegateTracer(DETECT_PROBE) === void 0 ? void 0 : provider.getDelegate();
	return provider;
}
function addSpanProcessor(provider, processor) {
	const providerWithProcessor = provider;
	if (typeof providerWithProcessor.addSpanProcessor === "function") {
		providerWithProcessor.addSpanProcessor(processor);
		return true;
	}
	const otelSpanProcessors = providerWithProcessor._activeSpanProcessor?._spanProcessors;
	if (Array.isArray(otelSpanProcessors)) {
		otelSpanProcessors.push(processor);
		return true;
	}
	const datadogSpanProcessors = providerWithProcessor._activeProcessor?._processors;
	if (Array.isArray(datadogSpanProcessors)) {
		datadogSpanProcessors.push(processor);
		return true;
	}
	return false;
}
var Latitude = class {
	provider;
	ready;
	latitudeProcessor;
	ownsProvider;
	constructor(options) {
		const { apiKey, project, projectSlug, instrumentations = {}, tracerProvider, ...processorOptionsRaw } = options;
		if (!apiKey || apiKey.trim() === "") throw new Error("[Latitude] apiKey is required and cannot be empty");
		if (project === void 0 && projectSlug !== void 0) warnProjectSlugDeprecated("constructor");
		const resolvedProject = project ?? projectSlug;
		const targetProvider = tracerProvider ?? getRegisteredTracerProvider();
		let processorOptions = processorOptionsRaw;
		if (targetProvider) {
			const { serviceName: _ignored, ...rest } = processorOptionsRaw;
			processorOptions = rest;
		}
		this.latitudeProcessor = new LatitudeSpanProcessor(apiKey, resolvedProject, processorOptions);
		const attached = targetProvider ? addSpanProcessor(targetProvider, this.latitudeProcessor) : false;
		if (targetProvider && !attached) console.warn(`[Latitude] Could not attach LatitudeSpanProcessor to ${tracerProvider ? "the provider passed via `tracerProvider`" : "the global OpenTelemetry provider"}: it exposes neither \`addSpanProcessor\` nor a known internal span-processor list (OTel JS v2 / Datadog). Falling back to a Latitude-owned provider that is NOT registered globally — instrumentations will still send spans to Latitude, but the host SDK's spans will not. To fix, pass a provider exposing \`addSpanProcessor\` or attach \`LatitudeSpanProcessor\` to your provider manually.`);
		if (targetProvider && attached) {
			this.provider = targetProvider;
			this.ownsProvider = false;
		} else {
			if (!targetProvider) {
				const contextManager = new AsyncLocalStorageContextManager();
				contextManager.enable();
				context.setGlobalContextManager(contextManager);
				propagation.setGlobalPropagator(new CompositePropagator({ propagators: [new W3CTraceContextPropagator(), new W3CBaggagePropagator()] }));
			}
			const resourceServiceName = typeof processorOptionsRaw.serviceName === "string" && processorOptionsRaw.serviceName.trim() !== "" ? processorOptionsRaw.serviceName.trim() : SERVICE_NAME;
			const latitudeProvider = new NodeTracerProvider({
				resource: resourceFromAttributes({ [ATTR_SERVICE_NAME]: resourceServiceName }),
				spanProcessors: [this.latitudeProcessor]
			});
			if (!targetProvider) latitudeProvider.register();
			this.ownsProvider = true;
			this.provider = latitudeProvider;
		}
		this.ready = registerLatitudeInstrumentations({
			instrumentations,
			tracerProvider: this.provider
		}).catch((err) => {
			console.warn("[Latitude] Failed to register instrumentations:", err);
		});
		if (!shutdownHandlersRegistered) {
			process.once("SIGTERM", () => void this.handleShutdown());
			process.once("SIGINT", () => void this.handleShutdown());
			shutdownHandlersRegistered = true;
		}
	}
	async shutdown() {
		if (this.ownsProvider && this.provider instanceof NodeTracerProvider) {
			await this.provider.shutdown();
			return;
		}
		await this.latitudeProcessor.shutdown();
	}
	async flush() {
		if (this.ownsProvider && this.provider instanceof NodeTracerProvider) {
			await this.provider.forceFlush();
			return;
		}
		await this.latitudeProcessor.forceFlush();
	}
	async handleShutdown() {
		try {
			await this.shutdown();
		} catch (err) {
			console.error("Error during Latitude Telemetry shutdown:", err);
		}
	}
};
/**
* @deprecated Use `new Latitude(options)` instead.
*/
function initLatitude(options) {
	return new Latitude(options);
}
//#endregion
export { ExportFilterSpanProcessor, Latitude, LatitudeSpanProcessor, RedactSpanProcessor, RedactThenExportSpanProcessor, buildShouldExportSpan, buildShouldExportSpanFromFields, capture, getLatitudeTracer, initLatitude, isDefaultExportSpan, isGenAiOrLlmAttributeSpan, isLatitudeInstrumentationSpan, registerLatitudeInstrumentations };
