// @latitude-data/telemetry/dist/instrumentation-DFz-ac4w.js
import { i as deriveFinishReason, r as buildOutputMessages, t as buildInputMessages } from "./messages-5-b5tIgg.js";
import { SpanKind, SpanStatusCode, context, trace } from "@opentelemetry/api";
import { ATTR_GEN_AI_INPUT_MESSAGES, ATTR_GEN_AI_OPERATION_NAME, ATTR_GEN_AI_OUTPUT_MESSAGES, ATTR_GEN_AI_PROVIDER_NAME, ATTR_GEN_AI_REQUEST_MAX_TOKENS, ATTR_GEN_AI_REQUEST_MODEL, ATTR_GEN_AI_REQUEST_TEMPERATURE, ATTR_GEN_AI_REQUEST_TOP_P, ATTR_GEN_AI_RESPONSE_FINISH_REASONS, ATTR_GEN_AI_RESPONSE_ID, ATTR_GEN_AI_RESPONSE_MODEL, ATTR_GEN_AI_USAGE_INPUT_TOKENS, ATTR_GEN_AI_USAGE_OUTPUT_TOKENS } from "@opentelemetry/semantic-conventions/incubating";
import { OpenAIInstrumentation } from "@traceloop/instrumentation-openai";
//#region src/sdk/instrumentations/openai/instrumentation.ts
/**
* Augments traceloop's `@traceloop/instrumentation-openai` with the OpenAI
* Responses API (`client.responses.create`), which traceloop's JS package does
* not yet patch. Mirrors the attribute conventions of the upstream Python
* `responses_wrappers.py`, so spans match what traceloop emits server-side.
*
* Once openllmetry-js ports their Python Responses support, this subclass
* should be deleted and the openai entry should switch back to the upstream
* `OpenAIInstrumentation` directly.
*/
const SPAN_NAME_PREFIX = "openai.response";
/**
* Drop-in replacement for `OpenAIInstrumentation` that also patches the
* Responses API on top of upstream Chat/Completions/Images coverage.
*/
var OpenAIInstrumentationWithResponses = class extends OpenAIInstrumentation {
	manuallyInstrument(module) {
		super.manuallyInstrument(module);
		const responsesProto = module?.Responses?.prototype;
		if (!responsesProto) return;
		this._wrap.call(this, responsesProto, "create", this.patchResponses());
	}
	patchResponses() {
		const plugin = this;
		return (original) => function patched(...args) {
			const params = args[0] ?? {};
			const span = plugin.tracer.startSpan(`${SPAN_NAME_PREFIX} ${params.model ?? "openai"}`, {
				kind: SpanKind.CLIENT,
				attributes: buildRequestAttributes(params)
			});
			const ctx = trace.setSpan(context.active(), span);
			return context.with(ctx, () => {
				let raw;
				try {
					raw = original.apply(this, args);
				} catch (err) {
					recordError(span, err);
					span.end();
					throw err;
				}
				const promise = raw;
				if (params.stream) return Promise.resolve(promise).then((stream) => wrapStream(span, stream), (err) => {
					recordError(span, err);
					span.end();
					throw err;
				});
				return Promise.resolve(promise).then((response) => {
					applyResponseAttributes(span, response);
					span.end();
					return response;
				}, (err) => {
					recordError(span, err);
					span.end();
					throw err;
				});
			});
		};
	}
};
function buildRequestAttributes(params) {
	const attrs = {
		[ATTR_GEN_AI_OPERATION_NAME]: "chat",
		[ATTR_GEN_AI_PROVIDER_NAME]: "openai"
	};
	if (params.model) attrs[ATTR_GEN_AI_REQUEST_MODEL] = params.model;
	if (params.max_output_tokens !== void 0) attrs[ATTR_GEN_AI_REQUEST_MAX_TOKENS] = params.max_output_tokens;
	if (params.temperature !== void 0) attrs[ATTR_GEN_AI_REQUEST_TEMPERATURE] = params.temperature;
	if (params.top_p !== void 0) attrs[ATTR_GEN_AI_REQUEST_TOP_P] = params.top_p;
	attrs[ATTR_GEN_AI_INPUT_MESSAGES] = JSON.stringify(buildInputMessages(params));
	return attrs;
}
function applyResponseAttributes(span, response) {
	if (!response || typeof response !== "object") return;
	if (response.id) span.setAttribute(ATTR_GEN_AI_RESPONSE_ID, response.id);
	if (response.model) span.setAttribute(ATTR_GEN_AI_RESPONSE_MODEL, response.model);
	const usage = response.usage;
	if (usage) {
		if (typeof usage.input_tokens === "number") span.setAttribute(ATTR_GEN_AI_USAGE_INPUT_TOKENS, usage.input_tokens);
		if (typeof usage.output_tokens === "number") span.setAttribute(ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, usage.output_tokens);
	}
	const finish = deriveFinishReason(response);
	if (finish) span.setAttribute(ATTR_GEN_AI_RESPONSE_FINISH_REASONS, [finish]);
	span.setAttribute(ATTR_GEN_AI_OUTPUT_MESSAGES, JSON.stringify(buildOutputMessages(response)));
}
function recordError(span, err) {
	span.setStatus({
		code: SpanStatusCode.ERROR,
		message: err instanceof Error ? err.message : String(err)
	});
	if (err instanceof Error) span.recordException(err);
}
/**
* Wrap an async iterable so we can capture the final `response.completed`
* payload (which carries usage and finalised output) before ending the span,
* while still forwarding every chunk to the caller transparently.
*/
async function* wrapStream(span, stream) {
	let finalResponse;
	try {
		for await (const event of stream) {
			const evt = event;
			if (evt?.type === "response.completed" && evt.response) finalResponse = evt.response;
			yield event;
		}
	} catch (err) {
		recordError(span, err);
		throw err;
	} finally {
		if (finalResponse) applyResponseAttributes(span, finalResponse);
		span.end();
	}
}
//#endregion
export { OpenAIInstrumentationWithResponses };
