// @latitude-data/telemetry/dist/messages-BfcVqxJE.cjs
//#region src/sdk/instrumentations/openai/messages.ts
function buildInputMessages(params) {
	const messages = [];
	if (params.instructions) messages.push({
		role: "system",
		parts: [{
			type: "text",
			content: params.instructions
		}]
	});
	if (typeof params.input === "string") messages.push({
		role: "user",
		parts: [{
			type: "text",
			content: params.input
		}]
	});
	else if (Array.isArray(params.input)) for (const item of params.input) messages.push(buildInputMessageFromItem(item));
	return messages;
}
function buildInputMessageFromItem(item) {
	const obj = item ?? {};
	const type = obj.type ?? "message";
	if (type === "function_call") {
		const callId = obj.callId ?? obj.call_id ?? obj.id;
		return {
			role: "assistant",
			parts: [{
				type: "tool_call",
				name: obj.name,
				id: callId,
				arguments: parseMaybeJson(obj.arguments)
			}]
		};
	}
	if (type === "function_call_output" || type === "function_call_result") return {
		role: "tool",
		parts: [{
			type: "tool_call_response",
			id: obj.callId ?? obj.call_id,
			response: extractToolResultOutput(obj.output)
		}]
	};
	const role = obj.role ?? "user";
	const content = obj.content;
	const parts = [];
	if (typeof content === "string") parts.push({
		type: "text",
		content
	});
	else if (Array.isArray(content)) for (const part of content) {
		const partObj = part ?? {};
		const partType = partObj.type ?? "";
		if (partType === "input_text" || partType === "output_text" || partType === "text") parts.push({
			type: "text",
			content: partObj.text ?? ""
		});
		else parts.push({
			type: partType || "unknown",
			content: partObj
		});
	}
	return {
		role,
		parts
	};
}
function extractToolResultOutput(output) {
	if (output === void 0 || output === null) return "";
	if (typeof output === "string") return output;
	if (typeof output === "object") {
		const out = output;
		if (typeof out.text === "string") return out.text;
	}
	return output;
}
function parseMaybeJson(value) {
	if (typeof value !== "string") return value;
	try {
		return JSON.parse(value);
	} catch {
		return value;
	}
}
function deriveFinishReason(response) {
	const status = response.status;
	if (status === "completed") return response.output?.some((b) => b.type === "function_call" || b.type === "file_search_call" || b.type === "web_search_call" || b.type === "computer_call") ? "tool_call" : "stop";
	if (status === "incomplete") return response.incomplete_details?.reason === "content_filter" ? "content_filter" : "length";
	if (status === "failed" || status === "cancelled") return "error";
	return "";
}
function buildOutputMessages(response) {
	const parts = [];
	const text = resolveOutputText(response);
	if (text) parts.push({
		type: "text",
		content: text
	});
	for (const block of response.output ?? []) if (block.type === "function_call") parts.push({
		type: "tool_call",
		name: block.name,
		id: block.id,
		arguments: parseMaybeJson(block.arguments)
	});
	if (parts.length === 0) return [];
	return [{
		role: "assistant",
		parts,
		finish_reason: deriveFinishReason(response)
	}];
}
/**
* Returns assistant text from a Responses-API response. Non-streaming responses
* carry a top-level `output_text` shortcut populated by the SDK's parser, but
* the streaming `response.completed` event payload does not — there we derive
* the text by concatenating `output_text` items inside `output[]` message blocks
* (mirroring what OpenAI's `addOutputText` parser does on the non-streaming path).
*/
function resolveOutputText(response) {
	if (response.output_text) return response.output_text;
	const chunks = [];
	for (const block of response.output ?? []) {
		if (block.type !== "message") continue;
		const content = block.content;
		if (!Array.isArray(content)) continue;
		for (const item of content) if (item?.type === "output_text" && typeof item.text === "string") chunks.push(item.text);
	}
	return chunks.join("");
}
/**
* Convert a list of OpenAI message-shaped items (Chat Completions / Agents
* `generation` payload) into the OTel GenAI parts shape.
*/
function buildMessagesFromList(items) {
	return items.map(buildInputMessageFromItem);
}
//#endregion
Object.defineProperty(exports, "buildInputMessages", {
	enumerable: true,
	get: function() {
		return buildInputMessages;
	}
});
Object.defineProperty(exports, "buildMessagesFromList", {
	enumerable: true,
	get: function() {
		return buildMessagesFromList;
	}
});
Object.defineProperty(exports, "buildOutputMessages", {
	enumerable: true,
	get: function() {
		return buildOutputMessages;
	}
});
Object.defineProperty(exports, "deriveFinishReason", {
	enumerable: true,
	get: function() {
		return deriveFinishReason;
	}
});
