// @traceloop/instrumentation-llamaindex/dist/index.mjs
import { safeExecuteInTheMiddle, InstrumentationBase, InstrumentationNodeModuleDefinition } from '@opentelemetry/instrumentation';
import { context, trace, SpanStatusCode, SpanKind } from '@opentelemetry/api';
import { SpanAttributes, TraceloopSpanKindValues, CONTEXT_KEY_ALLOW_TRACE_CONTENT, FinishReasons } from '@traceloop/ai-semantic-conventions';
import { ATTR_GEN_AI_PROVIDER_NAME, GEN_AI_PROVIDER_NAME_VALUE_OPENAI, ATTR_GEN_AI_REQUEST_MODEL, ATTR_GEN_AI_OPERATION_NAME, GEN_AI_OPERATION_NAME_VALUE_CHAT, ATTR_GEN_AI_REQUEST_TOP_P, ATTR_GEN_AI_REQUEST_TEMPERATURE, ATTR_GEN_AI_INPUT_MESSAGES, ATTR_GEN_AI_RESPONSE_MODEL, ATTR_GEN_AI_RESPONSE_ID, ATTR_GEN_AI_RESPONSE_FINISH_REASONS, ATTR_GEN_AI_USAGE_INPUT_TOKENS, ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, ATTR_GEN_AI_OUTPUT_MESSAGES } from '@opentelemetry/semantic-conventions/incubating';
import { formatInputMessages, mapOpenAIContentBlock, formatOutputMessage } from '@traceloop/instrumentation-utils';
import { __asyncGenerator, __asyncValues, __await } from 'tslib';
import * as lodash from 'lodash';

const shouldSendPrompts = (config) => {
    const contextShouldSendPrompts = context
        .active()
        .getValue(CONTEXT_KEY_ALLOW_TRACE_CONTENT);
    if (contextShouldSendPrompts !== undefined) {
        return !!contextShouldSendPrompts;
    }
    return config.traceContent !== undefined ? config.traceContent : true;
};
// Adopted from https://github.com/open-telemetry/opentelemetry-js/issues/2951#issuecomment-1214587378
function bindAsyncGenerator(ctx, generator) {
    var _a;
    return {
        next: context.bind(ctx, generator.next.bind(generator)),
        return: context.bind(ctx, generator.return.bind(generator)),
        throw: context.bind(ctx, generator.throw.bind(generator)),
        [Symbol.asyncIterator]() {
            return bindAsyncGenerator(ctx, generator[Symbol.asyncIterator]());
        },
        [Symbol.asyncDispose]: ((_a = generator[Symbol.asyncDispose]) === null || _a === void 0 ? void 0 : _a.bind(generator)) ||
            (() => Promise.resolve()),
    };
}
function generatorWrapper(streamingResult, ctx, fn) {
    return __asyncGenerator(this, arguments, function* generatorWrapper_1() {
        var _a, e_1, _b, _c;
        try {
            for (var _d = true, _e = __asyncValues(bindAsyncGenerator(ctx, streamingResult)), _f; _f = yield __await(_e.next()), _a = _f.done, !_a; _d = true) {
                _c = _f.value;
                _d = false;
                const chunk = _c;
                yield yield __await(chunk);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
            }
            finally { if (e_1) throw e_1.error; }
        }
        fn();
    });
}
function llmGeneratorWrapper(streamingResult, ctx, fn) {
    return __asyncGenerator(this, arguments, function* llmGeneratorWrapper_1() {
        var _a, e_2, _b, _c;
        let message = "";
        // Track the last chunk so the callback can extract usage/finish_reason from
        // chunk.raw — OpenAI sends these in the final streaming chunk when
        // stream_options: { include_usage: true } is set on the LLM.
        let lastChunk;
        let fnCalled = false;
        try {
            try {
                for (var _d = true, _e = __asyncValues(bindAsyncGenerator(ctx, streamingResult)), _f; _f = yield __await(_e.next()), _a = _f.done, !_a; _d = true) {
                    _c = _f.value;
                    _d = false;
                    const messageChunk = _c;
                    if (messageChunk.delta) {
                        message += messageChunk.delta;
                    }
                    if (messageChunk.text) {
                        message += messageChunk.text;
                    }
                    lastChunk = messageChunk;
                    yield yield __await(messageChunk);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        catch (err) {
            // Ensure span is finalized even if the stream throws
            if (!fnCalled) {
                fnCalled = true;
                fn(message, lastChunk);
            }
            throw err;
        }
        finally {
            // Covers normal completion and early consumer exit (break/return)
            if (!fnCalled) {
                fnCalled = true;
                fn(message, lastChunk);
            }
        }
    });
}
function genericWrapper(className, methodName, kind, tracer, shouldSendPrompts) {
    // eslint-disable-next-line
    return (original) => {
        return function method(...args) {
            const params = args[0];
            const streaming = params && params.stream;
            const name = `${lodash.snakeCase(className)}.${lodash.snakeCase(methodName)}`;
            const span = tracer().startSpan(`${name}`, {}, context.active());
            span.setAttribute(SpanAttributes.TRACELOOP_SPAN_KIND, kind);
            if (kind === TraceloopSpanKindValues.WORKFLOW) {
                span.setAttribute(SpanAttributes.TRACELOOP_WORKFLOW_NAME, name);
            }
            if (shouldSendPrompts) {
                try {
                    if (args.length === 1 &&
                        typeof args[0] === "object" &&
                        !(args[0] instanceof Map)) {
                        span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_INPUT, JSON.stringify({ args: [], kwargs: args[0] }));
                    }
                    else {
                        span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_INPUT, JSON.stringify({
                            args: args.map((arg) => arg instanceof Map ? Array.from(arg.entries()) : arg),
                            kwargs: {},
                        }));
                    }
                }
                catch (_a) {
                    /* empty */
                }
            }
            const execContext = trace.setSpan(context.active(), span);
            const execPromise = safeExecuteInTheMiddle(() => {
                return context.with(execContext, () => {
                    return original.apply(this, args);
                });
            }, 
            // eslint-disable-next-line @typescript-eslint/no-empty-function
            () => { });
            const wrappedPromise = execPromise
                .then((result) => {
                return new Promise((resolve) => {
                    if (streaming) {
                        result = generatorWrapper(result, execContext, () => {
                            span.setStatus({ code: SpanStatusCode.OK });
                            span.end();
                        });
                        resolve(result);
                    }
                    else {
                        span.setStatus({ code: SpanStatusCode.OK });
                        try {
                            if (shouldSendPrompts) {
                                if (result instanceof Map) {
                                    span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_OUTPUT, JSON.stringify(Array.from(result.entries())));
                                }
                                else {
                                    span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_OUTPUT, JSON.stringify(result));
                                }
                            }
                        }
                        finally {
                            span.end();
                            resolve(result);
                        }
                    }
                });
            })
                .catch((error) => {
                return new Promise((_, reject) => {
                    span.setStatus({
                        code: SpanStatusCode.ERROR,
                        message: error.message,
                    });
                    span.end();
                    reject(error);
                });
            });
            return context.bind(execContext, wrappedPromise);
        };
    };
}

const classNameToProviderName = {
    OpenAI: GEN_AI_PROVIDER_NAME_VALUE_OPENAI,
    // Future providers: Anthropic: "anthropic", Gemini: "gcp.gemini", etc.
    // See well-known values: https://opentelemetry.io/docs/specs/semconv/registry/attributes/gen-ai/#gen-ai-provider-name
};
const openAIFinishReasonMap = {
    stop: FinishReasons.STOP,
    length: FinishReasons.LENGTH,
    tool_calls: FinishReasons.TOOL_CALL,
    content_filter: FinishReasons.CONTENT_FILTER,
    function_call: FinishReasons.TOOL_CALL,
};
class CustomLLMInstrumentation {
    constructor(config, diag, tracer) {
        this.config = config;
        this.diag = diag;
        this.tracer = tracer;
    }
    chatWrapper({ className }) {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const plugin = this;
        return (original) => {
            return function method(...args) {
                var _a, _b, _c;
                const params = args[0];
                const messages = params === null || params === void 0 ? void 0 : params.messages;
                const streaming = params === null || params === void 0 ? void 0 : params.stream;
                const span = plugin.tracer().startSpan(`chat ${this.metadata.model}`, {
                    kind: SpanKind.CLIENT,
                });
                try {
                    span.setAttribute(ATTR_GEN_AI_PROVIDER_NAME, (_a = classNameToProviderName[className]) !== null && _a !== void 0 ? _a : className.toLowerCase());
                    span.setAttribute(ATTR_GEN_AI_REQUEST_MODEL, this.metadata.model);
                    span.setAttribute(ATTR_GEN_AI_OPERATION_NAME, GEN_AI_OPERATION_NAME_VALUE_CHAT);
                    span.setAttribute(ATTR_GEN_AI_REQUEST_TOP_P, this.metadata.topP);
                    span.setAttribute(ATTR_GEN_AI_REQUEST_TEMPERATURE, this.metadata.temperature);
                    if (shouldSendPrompts(plugin.config()) && messages) {
                        span.setAttribute(ATTR_GEN_AI_INPUT_MESSAGES, formatInputMessages(messages, mapOpenAIContentBlock));
                    }
                }
                catch (e) {
                    plugin.diag.warn(e);
                    (_c = (_b = plugin.config()).exceptionLogger) === null || _c === void 0 ? void 0 : _c.call(_b, e);
                }
                const execContext = trace.setSpan(context.active(), span);
                const execPromise = safeExecuteInTheMiddle(() => {
                    return context.with(execContext, () => {
                        return original.apply(this, args);
                    });
                }, 
                // eslint-disable-next-line @typescript-eslint/no-empty-function
                () => { });
                const wrappedPromise = execPromise
                    .then((result) => {
                    return new Promise((resolve) => {
                        if (streaming) {
                            result = plugin.handleStreamingResponse(result, span, execContext, this.metadata);
                        }
                        else {
                            result = plugin.handleResponse(result, span, this.metadata);
                        }
                        resolve(result);
                    });
                })
                    .catch((error) => {
                    return new Promise((_, reject) => {
                        span.setStatus({
                            code: SpanStatusCode.ERROR,
                            message: error.message,
                        });
                        span.end();
                        reject(error);
                    });
                });
                return context.bind(execContext, wrappedPromise);
            };
        };
    }
    handleResponse(result, span, metadata) {
        var _a, _b, _c, _d, _e, _f;
        span.setAttribute(ATTR_GEN_AI_RESPONSE_MODEL, metadata.model);
        try {
            const raw = result.raw;
            if (raw === null || raw === void 0 ? void 0 : raw.id) {
                span.setAttribute(ATTR_GEN_AI_RESPONSE_ID, raw.id);
            }
            const finishReason = (_c = (_b = (_a = raw === null || raw === void 0 ? void 0 : raw.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.finish_reason) !== null && _c !== void 0 ? _c : null;
            // finish_reasons: metadata, not content — always set outside shouldSendPrompts
            if (finishReason != null) {
                span.setAttribute(ATTR_GEN_AI_RESPONSE_FINISH_REASONS, [
                    (_d = openAIFinishReasonMap[finishReason]) !== null && _d !== void 0 ? _d : finishReason,
                ]);
            }
            // Token usage: always set when available
            const usage = raw === null || raw === void 0 ? void 0 : raw.usage;
            if (usage) {
                span.setAttribute(ATTR_GEN_AI_USAGE_INPUT_TOKENS, usage.prompt_tokens);
                span.setAttribute(ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, usage.completion_tokens);
                span.setAttribute(SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS, usage.total_tokens);
            }
            // output messages: content — always set inside shouldSendPrompts
            if (shouldSendPrompts(this.config()) &&
                result.message) {
                const content = result.message.content;
                // Normalize to array so mapOpenAIContentBlock handles both string and block array
                const contentArray = typeof content === "string" ? [content] : content;
                span.setAttribute(ATTR_GEN_AI_OUTPUT_MESSAGES, formatOutputMessage(contentArray, finishReason, openAIFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_CHAT, mapOpenAIContentBlock));
            }
            span.setStatus({ code: SpanStatusCode.OK });
        }
        catch (e) {
            this.diag.warn(e);
            (_f = (_e = this.config()).exceptionLogger) === null || _f === void 0 ? void 0 : _f.call(_e, e);
        }
        span.end();
        return result;
    }
    handleStreamingResponse(result, span, execContext, metadata) {
        span.setAttribute(ATTR_GEN_AI_RESPONSE_MODEL, metadata.model);
        return llmGeneratorWrapper(result, execContext, (message, lastChunk) => {
            var _a, _b, _c, _d, _e, _f, _g;
            try {
                // Extract finish_reason and usage from the last chunk's raw OpenAI
                // response — available when stream_options: { include_usage: true }
                // is set on the LLM (OpenAI sends usage in the final streaming chunk).
                const lastRaw = lastChunk === null || lastChunk === void 0 ? void 0 : lastChunk.raw;
                if (lastRaw === null || lastRaw === void 0 ? void 0 : lastRaw.id) {
                    span.setAttribute(ATTR_GEN_AI_RESPONSE_ID, lastRaw.id);
                }
                const finishReason = (_c = (_b = (_a = lastRaw === null || lastRaw === void 0 ? void 0 : lastRaw.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.finish_reason) !== null && _c !== void 0 ? _c : null;
                const usage = (_d = lastRaw === null || lastRaw === void 0 ? void 0 : lastRaw.usage) !== null && _d !== void 0 ? _d : null;
                if (finishReason != null) {
                    span.setAttribute(ATTR_GEN_AI_RESPONSE_FINISH_REASONS, [
                        (_e = openAIFinishReasonMap[finishReason]) !== null && _e !== void 0 ? _e : finishReason,
                    ]);
                }
                if (usage) {
                    span.setAttribute(ATTR_GEN_AI_USAGE_INPUT_TOKENS, usage.prompt_tokens);
                    span.setAttribute(ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, usage.completion_tokens);
                    span.setAttribute(SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS, usage.total_tokens);
                }
                if (!finishReason && !usage) {
                    this.diag.debug("LlamaIndex streaming: no finish_reason or usage in last chunk. " +
                        "Set stream_options: { include_usage: true } on the LLM to capture token usage.");
                }
                // Note: streaming only produces text parts — LlamaIndex's streaming interface
                // yields text deltas only, not full content blocks. Tool calls or multi-modal
                // content are collapsed into a single text string by llmGeneratorWrapper.
                if (shouldSendPrompts(this.config())) {
                    span.setAttribute(ATTR_GEN_AI_OUTPUT_MESSAGES, formatOutputMessage([message], finishReason, openAIFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_CHAT, mapOpenAIContentBlock));
                }
            }
            catch (e) {
                this.diag.warn(e);
                (_g = (_f = this.config()).exceptionLogger) === null || _g === void 0 ? void 0 : _g.call(_f, e);
            }
            span.setStatus({ code: SpanStatusCode.OK });
            span.end();
        });
    }
}

var version = "0.25.0";

class LlamaIndexInstrumentation extends InstrumentationBase {
    constructor(config = {}) {
        super("@traceloop/instrumentation-llamaindex", version, config);
        this.customLLMInstrumentation = new CustomLLMInstrumentation(() => this._config, this._diag, () => this.tracer);
    }
    setConfig(config = {}) {
        super.setConfig(config);
    }
    manuallyInstrument(module, openaiModule) {
        this._diag.debug("Manually instrumenting llamaindex");
        this.patch(module);
        if (openaiModule) {
            this.patchOpenAI(openaiModule);
        }
    }
    init() {
        const llamaindexModule = new InstrumentationNodeModuleDefinition("llamaindex", [">=0.1.0"], this.patch.bind(this), this.unpatch.bind(this));
        const openaiModule = new InstrumentationNodeModuleDefinition("@llamaindex/openai", [">=0.1.0"], this.patchOpenAI.bind(this), this.unpatchOpenAI.bind(this));
        return [llamaindexModule, openaiModule];
    }
    isLLM(llm) {
        return (llm &&
            llm.complete !== undefined &&
            llm.chat !== undefined);
    }
    isEmbedding(embedding) {
        return !!(embedding === null || embedding === void 0 ? void 0 : embedding.getQueryEmbedding);
    }
    isSynthesizer(synthesizer) {
        return (synthesizer && synthesizer.synthesize !== undefined);
    }
    isRetriever(retriever) {
        return retriever && retriever.retrieve !== undefined;
    }
    patch(moduleExports, moduleVersion) {
        this._diag.debug(`Patching llamaindex@${moduleVersion}`);
        this._wrap(moduleExports.RetrieverQueryEngine.prototype, "query", genericWrapper(moduleExports.RetrieverQueryEngine.name, "query", TraceloopSpanKindValues.WORKFLOW, () => this.tracer, shouldSendPrompts(this._config)));
        this._wrap(moduleExports.ContextChatEngine.prototype, "chat", genericWrapper(moduleExports.ContextChatEngine.name, "chat", TraceloopSpanKindValues.WORKFLOW, () => this.tracer, shouldSendPrompts(this._config)));
        // OpenAIAgent has been moved to @llamaindex/openai package in newer versions
        // This instrumentation is handled separately
        for (const key in moduleExports) {
            const cls = moduleExports[key];
            if (this.isLLM(cls.prototype)) {
                this._wrap(cls.prototype, "chat", this.customLLMInstrumentation.chatWrapper({ className: cls.name }));
            }
            else if (this.isEmbedding(cls.prototype)) {
                this._wrap(cls.prototype, "getQueryEmbedding", genericWrapper(cls.name, "getQueryEmbedding", TraceloopSpanKindValues.TASK, () => this.tracer, shouldSendPrompts(this._config)));
            }
            else if (this.isSynthesizer(cls.prototype)) {
                this._wrap(cls.prototype, "synthesize", genericWrapper(cls.name, "synthesize", TraceloopSpanKindValues.TASK, () => this.tracer, shouldSendPrompts(this._config)));
            }
            else if (this.isRetriever(cls.prototype)) {
                this._wrap(cls.prototype, "retrieve", genericWrapper(cls.name, "retrieve", TraceloopSpanKindValues.TASK, () => this.tracer, shouldSendPrompts(this._config)));
            }
        }
        return moduleExports;
    }
    unpatch(moduleExports, moduleVersion) {
        this._diag.debug(`Unpatching llamaindex@${moduleVersion}`);
        this._unwrap(moduleExports.RetrieverQueryEngine.prototype, "query");
        for (const key in moduleExports) {
            const cls = moduleExports[key];
            if (this.isLLM(cls.prototype)) {
                this._unwrap(cls.prototype, "complete");
                this._unwrap(cls.prototype, "chat");
            }
            else if (this.isEmbedding(cls.prototype)) {
                this._unwrap(cls.prototype, "getQueryEmbedding");
            }
            else if (this.isSynthesizer(cls.prototype)) {
                this._unwrap(cls.prototype, "synthesize");
            }
            else if (this.isRetriever(cls.prototype)) {
                this._unwrap(cls.prototype, "retrieve");
            }
        }
        return moduleExports;
    }
    patchOpenAI(moduleExports, moduleVersion) {
        this._diag.debug(`Patching @llamaindex/openai@${moduleVersion}`);
        if (moduleExports.OpenAI && this.isLLM(moduleExports.OpenAI.prototype)) {
            this._wrap(moduleExports.OpenAI.prototype, "chat", this.customLLMInstrumentation.chatWrapper({
                className: moduleExports.OpenAI.name,
            }));
        }
        if (moduleExports.OpenAIAgent && moduleExports.OpenAIAgent.prototype) {
            this._wrap(moduleExports.OpenAIAgent.prototype, "chat", genericWrapper(moduleExports.OpenAIAgent.name, "agent", TraceloopSpanKindValues.AGENT, () => this.tracer, shouldSendPrompts(this._config)));
        }
        return moduleExports;
    }
    unpatchOpenAI(moduleExports, moduleVersion) {
        this._diag.debug(`Unpatching @llamaindex/openai@${moduleVersion}`);
        if (moduleExports.OpenAI && moduleExports.OpenAI.prototype) {
            this._unwrap(moduleExports.OpenAI.prototype, "chat");
        }
        if (moduleExports.OpenAIAgent && moduleExports.OpenAIAgent.prototype) {
            this._unwrap(moduleExports.OpenAIAgent.prototype, "chat");
        }
        return moduleExports;
    }
}

export { LlamaIndexInstrumentation };
