// @traceloop/instrumentation-bedrock/dist/index.mjs
import { __awaiter, __asyncValues } from 'tslib';
import { trace, context, SpanStatusCode, SpanKind } from '@opentelemetry/api';
import { InstrumentationBase, InstrumentationNodeModuleDefinition, safeExecuteInTheMiddle } from '@opentelemetry/instrumentation';
import { FinishReasons, SpanAttributes, CONTEXT_KEY_ALLOW_TRACE_CONTENT } from '@traceloop/ai-semantic-conventions';
import { GEN_AI_PROVIDER_NAME_VALUE_AWS_BEDROCK, ATTR_GEN_AI_REQUEST_MODEL, ATTR_GEN_AI_PROVIDER_NAME, ATTR_GEN_AI_OPERATION_NAME, ATTR_GEN_AI_RESPONSE_MODEL, ATTR_GEN_AI_USAGE_INPUT_TOKENS, ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, ATTR_GEN_AI_RESPONSE_FINISH_REASONS, ATTR_GEN_AI_OUTPUT_MESSAGES, ATTR_GEN_AI_REQUEST_MAX_TOKENS, ATTR_GEN_AI_REQUEST_TEMPERATURE, ATTR_GEN_AI_REQUEST_TOP_P, ATTR_GEN_AI_INPUT_MESSAGES, ATTR_GEN_AI_REQUEST_TOP_K, ATTR_GEN_AI_REQUEST_FREQUENCY_PENALTY, ATTR_GEN_AI_REQUEST_PRESENCE_PENALTY, GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION, GEN_AI_OPERATION_NAME_VALUE_CHAT, ATTR_GEN_AI_SYSTEM_INSTRUCTIONS } from '@opentelemetry/semantic-conventions/incubating';
import { formatOutputMessage, mapBedrockContentBlock, formatInputMessagesFromPrompt, formatInputMessages, formatSystemInstructions } from '@traceloop/instrumentation-utils';

var BedrockVendor;
(function (BedrockVendor) {
    BedrockVendor["AI21"] = "ai21";
    BedrockVendor["AMAZON"] = "amazon";
    BedrockVendor["ANTHROPIC"] = "anthropic";
    BedrockVendor["COHERE"] = "cohere";
    BedrockVendor["META"] = "meta";
})(BedrockVendor || (BedrockVendor = {}));

var version = "0.26.0";

const bedrockFinishReasonMap = {
    // AI21
    endoftext: FinishReasons.STOP,
    // Amazon Titan / Nova
    FINISH: FinishReasons.STOP,
    LENGTH: FinishReasons.LENGTH,
    CONTENT_FILTERED: FinishReasons.CONTENT_FILTER,
    // Anthropic
    end_turn: FinishReasons.STOP,
    max_tokens: FinishReasons.LENGTH,
    stop_sequence: FinishReasons.STOP,
    tool_use: FinishReasons.TOOL_CALL,
    // Cohere
    COMPLETE: FinishReasons.STOP,
    MAX_TOKENS: FinishReasons.LENGTH,
    ERROR: FinishReasons.ERROR,
    ERROR_TOXIC: FinishReasons.CONTENT_FILTER,
    // Meta
    stop: FinishReasons.STOP,
    length: FinishReasons.LENGTH,
};
class BedrockInstrumentation extends InstrumentationBase {
    constructor(config = {}) {
        super("@traceloop/instrumentation-bedrock", version, config);
    }
    setConfig(config = {}) {
        super.setConfig(config);
    }
    init() {
        const module = new InstrumentationNodeModuleDefinition("@aws-sdk/client-bedrock-runtime", [">=3.499.0"], this.wrap.bind(this), this.unwrap.bind(this));
        return module;
    }
    manuallyInstrument(module) {
        this._diag.debug(`Patching @aws-sdk/client-bedrock-runtime manually`);
        this._wrap(module.BedrockRuntimeClient.prototype, "send", this.wrapperMethod());
    }
    wrap(module, moduleVersion) {
        this._diag.debug(`Patching @aws-sdk/client-bedrock-runtime@${moduleVersion}`);
        this._wrap(module.BedrockRuntimeClient.prototype, "send", this.wrapperMethod());
        return module;
    }
    unwrap(module, moduleVersion) {
        this._diag.debug(`Unpatching @aws-sdk/client-bedrock-runtime@${moduleVersion}`);
        this._unwrap(module.BedrockRuntimeClient.prototype, "send");
    }
    wrapperMethod() {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const plugin = this;
        // eslint-disable-next-line
        return (original) => {
            return function method(...args) {
                var _a, _b, _c;
                const span = plugin._startSpan({
                    params: args[0],
                });
                const modelId = (_c = (_b = (_a = args[0]) === null || _a === void 0 ? void 0 : _a.input) === null || _b === void 0 ? void 0 : _b.modelId) !== null && _c !== void 0 ? _c : "";
                const execContext = trace.setSpan(context.active(), span);
                const execPromise = safeExecuteInTheMiddle(() => {
                    return context.with(execContext, () => {
                        return original.apply(this, args);
                    });
                }, (e) => {
                    if (e) {
                        plugin._diag.error(`Error in bedrock instrumentation`, e);
                    }
                });
                const wrappedPromise = plugin._wrapPromise(span, execPromise, modelId);
                return context.bind(execContext, wrappedPromise);
            };
        };
    }
    _wrapPromise(span, promise, modelId) {
        return promise
            .then((result) => __awaiter(this, void 0, void 0, function* () {
            yield this._endSpan({
                span,
                result: result,
                modelId,
            });
            return new Promise((resolve) => resolve(result));
        }))
            .catch((error) => {
            return new Promise((_, reject) => {
                span.setStatus({
                    code: SpanStatusCode.ERROR,
                    message: error.message,
                });
                span.recordException(error);
                span.end();
                reject(error);
            });
        });
    }
    _startSpan({ params, }) {
        var _a, _b;
        let attributes = {};
        let spanName = "bedrock.completion";
        try {
            const input = params.input;
            const { modelVendor, model } = this._extractVendorAndModel(input.modelId || "");
            attributes = {
                [ATTR_GEN_AI_PROVIDER_NAME]: GEN_AI_PROVIDER_NAME_VALUE_AWS_BEDROCK,
                [ATTR_GEN_AI_REQUEST_MODEL]: model,
            };
            if (typeof input.body === "string") {
                const requestBody = JSON.parse(input.body);
                const operationType = this._getOperationType(modelVendor, requestBody);
                spanName = `${operationType} ${model}`;
                // Set operation name before _setRequestAttributes so it is always
                // present even if _setRequestAttributes throws for an unexpected body.
                attributes = Object.assign(Object.assign({}, attributes), { [ATTR_GEN_AI_OPERATION_NAME]: operationType });
                attributes = Object.assign(Object.assign({}, attributes), this._setRequestAttributes(modelVendor, requestBody));
            }
        }
        catch (e) {
            this._diag.debug(e);
            (_b = (_a = this._config).exceptionLogger) === null || _b === void 0 ? void 0 : _b.call(_a, e);
        }
        return this.tracer.startSpan(spanName, {
            kind: SpanKind.CLIENT,
            attributes,
        });
    }
    _endSpan(_a) {
        return __awaiter(this, arguments, void 0, function* ({ span, result, modelId, }) {
            var _b, e_1, _c, _d;
            var _e, _f, _g, _h;
            try {
                if ("body" in result) {
                    const attributes = "attributes" in span
                        ? span["attributes"]
                        : {};
                    if (ATTR_GEN_AI_PROVIDER_NAME in attributes) {
                        const { modelVendor, model } = this._extractVendorAndModel(modelId);
                        span.setAttribute(ATTR_GEN_AI_RESPONSE_MODEL, model);
                        if (!(result.body instanceof Object.getPrototypeOf(Uint8Array))) {
                            const rawRes = result.body;
                            let streamedContent = "";
                            try {
                                for (var _j = true, rawRes_1 = __asyncValues(rawRes), rawRes_1_1; rawRes_1_1 = yield rawRes_1.next(), _b = rawRes_1_1.done, !_b; _j = true) {
                                    _d = rawRes_1_1.value;
                                    _j = false;
                                    const value = _d;
                                    // Convert it to a JSON String
                                    const jsonString = new TextDecoder().decode((_e = value.chunk) === null || _e === void 0 ? void 0 : _e.bytes);
                                    // Parse the JSON string
                                    const parsedResponse = JSON.parse(jsonString);
                                    if ("amazon-bedrock-invocationMetrics" in parsedResponse) {
                                        span.setAttribute(ATTR_GEN_AI_USAGE_INPUT_TOKENS, parsedResponse["amazon-bedrock-invocationMetrics"]["inputTokenCount"]);
                                        span.setAttribute(ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, parsedResponse["amazon-bedrock-invocationMetrics"]["outputTokenCount"]);
                                        const totalTokens = parsedResponse["amazon-bedrock-invocationMetrics"]["inputTokenCount"] +
                                            parsedResponse["amazon-bedrock-invocationMetrics"]["outputTokenCount"];
                                        span.setAttribute(SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS, totalTokens);
                                    }
                                    this._handleNovaStreamingMetadata(span, parsedResponse);
                                    const responseAttributes = this._setResponseAttributes(modelVendor, parsedResponse, true);
                                    // ! NOTE: This make sure the content always have all streamed chunks
                                    if (this._shouldSendPrompts()) {
                                        const chunkContent = this._getStreamChunkContent(modelVendor, parsedResponse);
                                        if (chunkContent !== undefined) {
                                            streamedContent += chunkContent;
                                        }
                                        // When finish reason is available (final chunk), set OTel 1.40 output message
                                        if (responseAttributes[ATTR_GEN_AI_RESPONSE_FINISH_REASONS]) {
                                            const finishReasons = responseAttributes[ATTR_GEN_AI_RESPONSE_FINISH_REASONS];
                                            responseAttributes[ATTR_GEN_AI_OUTPUT_MESSAGES] =
                                                formatOutputMessage(streamedContent, (_f = finishReasons[0]) !== null && _f !== void 0 ? _f : null, {}, // already mapped by _setResponseAttributes
                                                attributes[ATTR_GEN_AI_OPERATION_NAME], mapBedrockContentBlock);
                                        }
                                    }
                                    span.setAttributes(responseAttributes);
                                }
                            }
                            catch (e_1_1) { e_1 = { error: e_1_1 }; }
                            finally {
                                try {
                                    if (!_j && !_b && (_c = rawRes_1.return)) yield _c.call(rawRes_1);
                                }
                                finally { if (e_1) throw e_1.error; }
                            }
                        }
                        else if (result.body instanceof Object.getPrototypeOf(Uint8Array)) {
                            // Convert it to a JSON String
                            const jsonString = new TextDecoder().decode(result.body);
                            // Parse the JSON string
                            const parsedResponse = JSON.parse(jsonString);
                            const responseAttributes = this._setResponseAttributes(modelVendor, parsedResponse);
                            span.setAttributes(responseAttributes);
                        }
                    }
                }
            }
            catch (e) {
                this._diag.debug(e);
                (_h = (_g = this._config).exceptionLogger) === null || _h === void 0 ? void 0 : _h.call(_g, e);
            }
            span.setStatus({ code: SpanStatusCode.OK });
            span.end();
        });
    }
    _setRequestAttributes(vendor, requestBody) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        switch (vendor) {
            case BedrockVendor.AI21: {
                // Jamba format: messages array + max_tokens + top_p
                if (requestBody["messages"]) {
                    return Object.assign({ [ATTR_GEN_AI_REQUEST_TOP_P]: requestBody["top_p"], [ATTR_GEN_AI_REQUEST_TEMPERATURE]: requestBody["temperature"], [ATTR_GEN_AI_REQUEST_MAX_TOKENS]: requestBody["max_tokens"] }, (this._shouldSendPrompts()
                        ? {
                            [ATTR_GEN_AI_INPUT_MESSAGES]: formatInputMessages(requestBody["messages"], mapBedrockContentBlock),
                        }
                        : {}));
                }
                // Legacy Jurassic format: prompt + topP + maxTokens
                return Object.assign({ [ATTR_GEN_AI_REQUEST_TOP_P]: requestBody["topP"], [ATTR_GEN_AI_REQUEST_TEMPERATURE]: requestBody["temperature"], [ATTR_GEN_AI_REQUEST_MAX_TOKENS]: requestBody["maxTokens"], [ATTR_GEN_AI_REQUEST_PRESENCE_PENALTY]: (_a = requestBody["presencePenalty"]) === null || _a === void 0 ? void 0 : _a["scale"], [ATTR_GEN_AI_REQUEST_FREQUENCY_PENALTY]: (_b = requestBody["frequencyPenalty"]) === null || _b === void 0 ? void 0 : _b["scale"] }, (this._shouldSendPrompts()
                    ? {
                        [ATTR_GEN_AI_INPUT_MESSAGES]: formatInputMessagesFromPrompt(requestBody["prompt"]),
                    }
                    : {}));
            }
            case BedrockVendor.AMAZON: {
                // Amazon Nova format: messages array + inferenceConfig
                if (requestBody["messages"]) {
                    return Object.assign({ [ATTR_GEN_AI_REQUEST_TOP_P]: (_c = requestBody["inferenceConfig"]) === null || _c === void 0 ? void 0 : _c["topP"], [ATTR_GEN_AI_REQUEST_TEMPERATURE]: (_d = requestBody["inferenceConfig"]) === null || _d === void 0 ? void 0 : _d["temperature"], [ATTR_GEN_AI_REQUEST_MAX_TOKENS]: (_e = requestBody["inferenceConfig"]) === null || _e === void 0 ? void 0 : _e["maxTokens"] }, (this._shouldSendPrompts()
                        ? {
                            [ATTR_GEN_AI_INPUT_MESSAGES]: formatInputMessages(requestBody["messages"], mapBedrockContentBlock),
                        }
                        : {}));
                }
                // Amazon Titan format: inputText + textGenerationConfig
                return Object.assign({ [ATTR_GEN_AI_REQUEST_TOP_P]: (_f = requestBody["textGenerationConfig"]) === null || _f === void 0 ? void 0 : _f["topP"], [ATTR_GEN_AI_REQUEST_TEMPERATURE]: (_g = requestBody["textGenerationConfig"]) === null || _g === void 0 ? void 0 : _g["temperature"], [ATTR_GEN_AI_REQUEST_MAX_TOKENS]: (_h = requestBody["textGenerationConfig"]) === null || _h === void 0 ? void 0 : _h["maxTokenCount"] }, (this._shouldSendPrompts()
                    ? {
                        [ATTR_GEN_AI_INPUT_MESSAGES]: formatInputMessagesFromPrompt(requestBody["inputText"]),
                    }
                    : {}));
            }
            case BedrockVendor.ANTHROPIC: {
                const baseAttributes = {
                    [ATTR_GEN_AI_REQUEST_TOP_P]: requestBody["top_p"],
                    [ATTR_GEN_AI_REQUEST_TOP_K]: requestBody["top_k"],
                    [ATTR_GEN_AI_REQUEST_TEMPERATURE]: requestBody["temperature"],
                    [ATTR_GEN_AI_REQUEST_MAX_TOKENS]: requestBody["max_tokens_to_sample"] || requestBody["max_tokens"],
                };
                if (!this._shouldSendPrompts()) {
                    return baseAttributes;
                }
                // Handle new messages API format (used by langchain)
                if (requestBody["messages"]) {
                    const promptAttributes = {
                        [ATTR_GEN_AI_INPUT_MESSAGES]: formatInputMessages(requestBody["messages"], mapBedrockContentBlock),
                    };
                    if (requestBody["system"] !== undefined) {
                        promptAttributes[ATTR_GEN_AI_SYSTEM_INSTRUCTIONS] =
                            formatSystemInstructions(requestBody["system"]);
                    }
                    return Object.assign(Object.assign({}, baseAttributes), promptAttributes);
                }
                // Handle legacy prompt format
                if (requestBody["prompt"]) {
                    return Object.assign(Object.assign({}, baseAttributes), { [ATTR_GEN_AI_INPUT_MESSAGES]: formatInputMessagesFromPrompt(requestBody["prompt"]
                            .replace("\n\nHuman:", "")
                            .replace("\n\nAssistant:", "")) });
                }
                return baseAttributes;
            }
            case BedrockVendor.COHERE: {
                return {
                    [ATTR_GEN_AI_REQUEST_TOP_P]: requestBody["p"],
                    [ATTR_GEN_AI_REQUEST_TOP_K]: requestBody["k"],
                    [ATTR_GEN_AI_REQUEST_TEMPERATURE]: requestBody["temperature"],
                    [ATTR_GEN_AI_REQUEST_MAX_TOKENS]: requestBody["max_tokens"],
                };
            }
            case BedrockVendor.META: {
                return Object.assign({ [ATTR_GEN_AI_REQUEST_TOP_P]: requestBody["top_p"], [ATTR_GEN_AI_REQUEST_TEMPERATURE]: requestBody["temperature"], [ATTR_GEN_AI_REQUEST_MAX_TOKENS]: requestBody["max_gen_len"] }, (this._shouldSendPrompts()
                    ? {
                        [ATTR_GEN_AI_INPUT_MESSAGES]: formatInputMessagesFromPrompt(requestBody["prompt"]),
                    }
                    : {}));
            }
            default:
                return {};
        }
    }
    _setResponseAttributes(vendor, response, isStream = false) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8;
        switch (vendor) {
            case BedrockVendor.AI21: {
                // Jamba format: choices[0].message + choices[0].finish_reason
                if (response["choices"]) {
                    const usage = response["usage"];
                    const finishReason = (_a = response["choices"][0]) === null || _a === void 0 ? void 0 : _a["finish_reason"];
                    const content = isStream
                        ? (_c = (_b = response["choices"][0]) === null || _b === void 0 ? void 0 : _b["delta"]) === null || _c === void 0 ? void 0 : _c["content"]
                        : (_e = (_d = response["choices"][0]) === null || _d === void 0 ? void 0 : _d["message"]) === null || _e === void 0 ? void 0 : _e["content"];
                    return Object.assign(Object.assign(Object.assign({}, (finishReason != null
                        ? {
                            [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                                (_f = bedrockFinishReasonMap[finishReason]) !== null && _f !== void 0 ? _f : finishReason,
                            ],
                        }
                        : {})), (usage
                        ? {
                            [ATTR_GEN_AI_USAGE_INPUT_TOKENS]: usage["prompt_tokens"],
                            [ATTR_GEN_AI_USAGE_OUTPUT_TOKENS]: usage["completion_tokens"],
                            [SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS]: usage["total_tokens"],
                        }
                        : {})), (this._shouldSendPrompts()
                        ? {
                            [ATTR_GEN_AI_OUTPUT_MESSAGES]: formatOutputMessage(content !== null && content !== void 0 ? content : "", finishReason, bedrockFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION, mapBedrockContentBlock),
                        }
                        : {}));
                }
                // Legacy Jurassic format: completions[0].data.text
                const jurassicFinishReason = response["completions"][0]["finishReason"]["reason"];
                const jurassicContent = response["completions"][0]["data"]["text"];
                return Object.assign({ [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                        (_g = bedrockFinishReasonMap[jurassicFinishReason]) !== null && _g !== void 0 ? _g : jurassicFinishReason,
                    ] }, (this._shouldSendPrompts()
                    ? {
                        [ATTR_GEN_AI_OUTPUT_MESSAGES]: formatOutputMessage(jurassicContent, jurassicFinishReason, bedrockFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION, mapBedrockContentBlock),
                    }
                    : {}));
            }
            case BedrockVendor.AMAZON: {
                if (isStream) {
                    // Amazon Nova format: contentBlockDelta has text, messageStop has stopReason
                    const novaFinishReason = (_h = response["messageStop"]) === null || _h === void 0 ? void 0 : _h["stopReason"];
                    // Amazon Titan format: outputText, completionReason
                    const titanFinishReason = response["completionReason"];
                    const finishReason = novaFinishReason !== null && novaFinishReason !== void 0 ? novaFinishReason : titanFinishReason;
                    return Object.assign(Object.assign({}, (finishReason != null
                        ? {
                            [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                                (_j = bedrockFinishReasonMap[finishReason]) !== null && _j !== void 0 ? _j : finishReason,
                            ],
                        }
                        : {})), (response["inputTextTokenCount"] != null
                        ? {
                            [ATTR_GEN_AI_USAGE_INPUT_TOKENS]: response["inputTextTokenCount"],
                            [ATTR_GEN_AI_USAGE_OUTPUT_TOKENS]: response["totalOutputTextTokenCount"],
                            [SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS]: response["inputTextTokenCount"] +
                                response["totalOutputTextTokenCount"],
                        }
                        : {}));
                }
                // Amazon Titan token fields
                const titanInputTokens = response["inputTextTokenCount"];
                const titanOutputTokens = (_l = (_k = response["results"]) === null || _k === void 0 ? void 0 : _k[0]) === null || _l === void 0 ? void 0 : _l["tokenCount"];
                // Amazon Nova token fields
                const novaUsage = response["usage"];
                const amazonFinishReason = (_p = (_o = (_m = response["results"]) === null || _m === void 0 ? void 0 : _m[0]) === null || _o === void 0 ? void 0 : _o["completionReason"]) !== null && _p !== void 0 ? _p : response["stopReason"];
                const novaRawContent = (_r = (_q = response["output"]) === null || _q === void 0 ? void 0 : _q["message"]) === null || _r === void 0 ? void 0 : _r["content"];
                const titanContent = (_t = (_s = response["results"]) === null || _s === void 0 ? void 0 : _s[0]) === null || _t === void 0 ? void 0 : _t["outputText"];
                const outputContent = (_u = novaRawContent !== null && novaRawContent !== void 0 ? novaRawContent : titanContent) !== null && _u !== void 0 ? _u : "";
                const operationType = novaRawContent
                    ? GEN_AI_OPERATION_NAME_VALUE_CHAT
                    : GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION;
                return Object.assign(Object.assign(Object.assign({}, (amazonFinishReason != null
                    ? {
                        [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                            (_v = bedrockFinishReasonMap[amazonFinishReason]) !== null && _v !== void 0 ? _v : amazonFinishReason,
                        ],
                    }
                    : {})), (titanInputTokens != null
                    ? {
                        [ATTR_GEN_AI_USAGE_INPUT_TOKENS]: titanInputTokens,
                        [ATTR_GEN_AI_USAGE_OUTPUT_TOKENS]: titanOutputTokens,
                        [SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS]: titanInputTokens + titanOutputTokens,
                    }
                    : novaUsage != null
                        ? {
                            [ATTR_GEN_AI_USAGE_INPUT_TOKENS]: novaUsage["inputTokens"],
                            [ATTR_GEN_AI_USAGE_OUTPUT_TOKENS]: novaUsage["outputTokens"],
                            [SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS]: novaUsage["totalTokens"],
                        }
                        : {})), (this._shouldSendPrompts()
                    ? {
                        [ATTR_GEN_AI_OUTPUT_MESSAGES]: formatOutputMessage(outputContent, amazonFinishReason, bedrockFinishReasonMap, operationType, mapBedrockContentBlock),
                    }
                    : {}));
            }
            case BedrockVendor.ANTHROPIC: {
                if (isStream) {
                    // New messages API streaming: content_block_delta has delta.text,
                    // message_delta has delta.stop_reason
                    const stopReason = (_x = (_w = response["delta"]) === null || _w === void 0 ? void 0 : _w["stop_reason"]) !== null && _x !== void 0 ? _x : response["stop_reason"];
                    const finishReason = stopReason !== null && stopReason !== void 0 ? stopReason : undefined;
                    return Object.assign({}, (finishReason != null
                        ? {
                            [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                                (_y = bedrockFinishReasonMap[finishReason]) !== null && _y !== void 0 ? _y : finishReason,
                            ],
                        }
                        : {}));
                }
                const stopReason = response["stop_reason"];
                const usage = response["usage"];
                const baseAttributes = Object.assign(Object.assign({}, (stopReason != null
                    ? {
                        [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                            (_z = bedrockFinishReasonMap[stopReason]) !== null && _z !== void 0 ? _z : stopReason,
                        ],
                    }
                    : {})), (usage
                    ? {
                        [ATTR_GEN_AI_USAGE_INPUT_TOKENS]: usage["input_tokens"],
                        [ATTR_GEN_AI_USAGE_OUTPUT_TOKENS]: usage["output_tokens"],
                        [SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS]: (usage["input_tokens"] || 0) + (usage["output_tokens"] || 0),
                    }
                    : {}));
                if (!this._shouldSendPrompts()) {
                    return baseAttributes;
                }
                // Handle new messages API format response
                if (response["content"]) {
                    const content = Array.isArray(response["content"])
                        ? response["content"]
                        : response["content"];
                    return Object.assign(Object.assign({}, baseAttributes), { [ATTR_GEN_AI_OUTPUT_MESSAGES]: formatOutputMessage(content, stopReason, bedrockFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_CHAT, mapBedrockContentBlock) });
                }
                // Handle legacy completion format
                if (response["completion"]) {
                    return Object.assign(Object.assign({}, baseAttributes), { [ATTR_GEN_AI_OUTPUT_MESSAGES]: formatOutputMessage(response["completion"], stopReason, bedrockFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION, mapBedrockContentBlock) });
                }
                return baseAttributes;
            }
            case BedrockVendor.COHERE: {
                const cohereFinishReason = (_2 = (_1 = (_0 = response["generations"]) === null || _0 === void 0 ? void 0 : _0[0]) === null || _1 === void 0 ? void 0 : _1["finish_reason"]) !== null && _2 !== void 0 ? _2 : response["finish_reason"];
                const cohereText = (_5 = (_4 = (_3 = response["generations"]) === null || _3 === void 0 ? void 0 : _3[0]) === null || _4 === void 0 ? void 0 : _4["text"]) !== null && _5 !== void 0 ? _5 : response["text"];
                return Object.assign(Object.assign(Object.assign({}, (cohereFinishReason != null
                    ? {
                        [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                            (_6 = bedrockFinishReasonMap[cohereFinishReason]) !== null && _6 !== void 0 ? _6 : cohereFinishReason,
                        ],
                    }
                    : {})), (((_7 = response["meta"]) === null || _7 === void 0 ? void 0 : _7["billed_units"])
                    ? {
                        [ATTR_GEN_AI_USAGE_INPUT_TOKENS]: response["meta"]["billed_units"]["input_tokens"],
                        [ATTR_GEN_AI_USAGE_OUTPUT_TOKENS]: response["meta"]["billed_units"]["output_tokens"],
                        [SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS]: (response["meta"]["billed_units"]["input_tokens"] || 0) +
                            (response["meta"]["billed_units"]["output_tokens"] || 0),
                    }
                    : {})), (this._shouldSendPrompts() && cohereText != null
                    ? {
                        [ATTR_GEN_AI_OUTPUT_MESSAGES]: formatOutputMessage(cohereText, cohereFinishReason, bedrockFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION, mapBedrockContentBlock),
                    }
                    : {}));
            }
            case BedrockVendor.META: {
                const metaFinishReason = response["stop_reason"];
                const metaContent = response["generation"];
                return Object.assign(Object.assign(Object.assign({}, (metaFinishReason != null
                    ? {
                        [ATTR_GEN_AI_RESPONSE_FINISH_REASONS]: [
                            (_8 = bedrockFinishReasonMap[metaFinishReason]) !== null && _8 !== void 0 ? _8 : metaFinishReason,
                        ],
                    }
                    : {})), { [ATTR_GEN_AI_USAGE_INPUT_TOKENS]: response["prompt_token_count"], [ATTR_GEN_AI_USAGE_OUTPUT_TOKENS]: response["generation_token_count"], [SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS]: response["prompt_token_count"] + response["generation_token_count"] }), (this._shouldSendPrompts()
                    ? {
                        [ATTR_GEN_AI_OUTPUT_MESSAGES]: formatOutputMessage(metaContent !== null && metaContent !== void 0 ? metaContent : "", metaFinishReason, bedrockFinishReasonMap, GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION, mapBedrockContentBlock),
                    }
                    : {}));
            }
            default:
                return {};
        }
    }
    _handleNovaStreamingMetadata(span, parsedResponse) {
        var _a;
        if ("metadata" in parsedResponse && ((_a = parsedResponse["metadata"]) === null || _a === void 0 ? void 0 : _a["usage"])) {
            const usage = parsedResponse["metadata"]["usage"];
            span.setAttribute(ATTR_GEN_AI_USAGE_INPUT_TOKENS, usage["inputTokens"]);
            span.setAttribute(ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, usage["outputTokens"]);
            span.setAttribute(SpanAttributes.GEN_AI_USAGE_TOTAL_TOKENS, usage["totalTokens"]);
        }
    }
    _shouldSendPrompts() {
        const contextShouldSendPrompts = context
            .active()
            .getValue(CONTEXT_KEY_ALLOW_TRACE_CONTENT);
        if (contextShouldSendPrompts !== undefined) {
            return contextShouldSendPrompts;
        }
        return this._config.traceContent !== undefined
            ? this._config.traceContent
            : true;
    }
    _getOperationType(vendor, requestBody) {
        switch (vendor) {
            case BedrockVendor.AI21:
            case BedrockVendor.AMAZON:
            case BedrockVendor.ANTHROPIC:
                return requestBody["messages"]
                    ? GEN_AI_OPERATION_NAME_VALUE_CHAT
                    : GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION;
            case BedrockVendor.COHERE:
            case BedrockVendor.META:
                return GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION;
            default:
                return GEN_AI_OPERATION_NAME_VALUE_TEXT_COMPLETION;
        }
    }
    _getStreamChunkContent(vendor, response) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        switch (vendor) {
            case BedrockVendor.AMAZON:
                return ((_c = (_b = (_a = response["contentBlockDelta"]) === null || _a === void 0 ? void 0 : _a["delta"]) === null || _b === void 0 ? void 0 : _b["text"]) !== null && _c !== void 0 ? _c : response["outputText"]);
            case BedrockVendor.ANTHROPIC:
                return (_e = (_d = response["delta"]) === null || _d === void 0 ? void 0 : _d["text"]) !== null && _e !== void 0 ? _e : response["completion"];
            case BedrockVendor.AI21:
                return (_h = (_g = (_f = response["choices"]) === null || _f === void 0 ? void 0 : _f[0]) === null || _g === void 0 ? void 0 : _g["delta"]) === null || _h === void 0 ? void 0 : _h["content"];
            case BedrockVendor.META:
                return response["generation"];
            default:
                return undefined;
        }
    }
    _extractVendorAndModel(modelId) {
        if (!modelId) {
            return { modelVendor: "", model: "" };
        }
        // Handle cross-region inference profile IDs like "us.anthropic.claude-3-5-sonnet-20241022-v2:0"
        // Mirrors Python's _cross_region_check logic.
        const prefixes = ["us", "us-gov", "eu", "apac"];
        const hasCrossRegionPrefix = prefixes.some((prefix) => modelId.startsWith(prefix + "."));
        if (hasCrossRegionPrefix) {
            const parts = modelId.split(".");
            if (parts.length > 2) {
                parts.shift(); // remove region prefix ("us", "eu", etc.)
            }
            return { modelVendor: parts[0] || "", model: parts[1] || "" };
        }
        // Standard format: "vendor.model-name"
        const dotIndex = modelId.indexOf(".");
        if (dotIndex === -1) {
            return { modelVendor: modelId, model: "" };
        }
        return {
            modelVendor: modelId.slice(0, dotIndex),
            model: modelId.slice(dotIndex + 1),
        };
    }
}

export { BedrockInstrumentation, BedrockVendor, bedrockFinishReasonMap };
